{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path\n",
    "sys.path.append(\"src/\")\n",
    "from estimate_pop_parameters import *\n",
    "import calculate_theta_from_fastsimcoal_output as fscoal"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Estimating population parameters from UCE allele alignments\n",
    "Here I wrote some functions that help extracting useful summary statistics from UCE allele alignments and plotting them."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Getting all variable positions from allele alignments\n",
    "The underlying data are multiple sequence alignments containing the UCE allele sequences of 10 diploid individuals (=20 sequences). The data looks like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "ls ../data/topaza-uce-allele-alignments/ | head -n 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "head -n 60 ../data/topaza-uce-allele-alignments/uce-101_phased.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Before we continue we need to define a few variables:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "input_dir = '../data/topaza-uce-allele-alignments'\n",
    "aln_format = 'fasta'\n",
    "min_num_seqs = 20\n",
    "output_dir = '../results/plots'\n",
    "outgroup = 'Florisuga'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we want to know which positions in the alignments contain polymorphisms and have data for all sequences. The following function does exactly that:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# get a list of all alignments\n",
    "alignment_list = read_aln_in_folder(input_dir,aln_format)\n",
    "# return a dictionary of all positions with missing data and another one only containing variable positions without missing data\n",
    "valid_columns, snp_columns, removed_seqs = get_variable_positions(alignment_list,aln_format,min_num_seqs,outgroup)\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The resulting dictionary 'snp_columns' contains a list of all variable positions for each locus, here is a small excerpt:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "dict(list(snp_columns.items())[0:10])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From these variable positions (=SNPs) we eventually want to estimate theta, using Tajima's estimator. For this we need to compare every pair of sequences and count all differences and then divide by the total number of pairwise comparisons. Here we retrieve the total number of pairwise differences for our 20 sequences (minus all sequences which represent outgroup seqeunces and were removed by the previous function):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "possible_combos = get_possible_combinations_for_n_sequences(int(min_num_seqs)-int(removed_seqs))\n",
    "print(possible_combos)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "However, making all these pairwise comparisons can take an aweful long time. Here we use a shortcut, only comparing those positions which we have previously identified to be polymorphic and calculating how many pairwise differences would return a difference, based on the frequency of the SNP. Tajima's estimator is then calculated for each locus and corrected for the total alignment length."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "locus_tajima_dict,locus_snp_count_dict = tajimas_estimator_per_locus(input_dir,aln_format,snp_columns,valid_columns,min_num_seqs,outgroup,possible_combos)\n",
    "dict(list(locus_tajima_dict.items())[0:10])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's plot the calculated thetas for all loci and calculate the expected fraction of heterozygos sites within the ingroup:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "plot_expected_heterozyosity(locus_tajima_dict,output_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We are also interested in the site frequency spectrum (SFS) across all loci within our ingroup. We can use the outgroup in order to guess which allele is the original state and which is the derived one. The function follows the following logic:\n",
    "\n",
    "A) determine the ancestral state at each polymorphism:\n",
    "    1. if outgroup allele also present in ingroup, take it as ancestral state\n",
    "    2. if outgroup allele not present in ingroup, choose the most common ingroup allele as ancestral\n",
    "    3. if all allele copies are equally common, choose one at random as ancestral\n",
    "B) for all derived mutations, count by how many sequences those are shared"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "locus_derived_allele_counts_dict = get_fsf_stats(input_dir,aln_format,outgroup,locus_snp_count_dict)\n",
    "dict(list(locus_derived_allele_counts_dict.items())[0:10])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now join all these values for each single locus into one global list and plot it as a histogram:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "join_and_plot_sfs_data(locus_derived_allele_counts_dict,output_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Estimating theta from fastsimcoal sequence simulation output\n",
    "In the course we applied abc methods to estimate theta from simulated data. Here I calculate theta using the Tajima estimator instead:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tajima estimated log(theta) = -2.666712\n",
      "real log(theta) = -2.698970\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYsAAAEaCAYAAADg2nttAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzs3Xd8FHX6wPHPkwKhd5DeW0IIJYAKUkXwbGAFQUEPEA+s\nd5b76Z2enp5nQ1FPwC7qWVAQAc8KKnZApCm9N0NJII2UfX5/zCRulpRNyGazyfN+vfa1O+07z8zO\nzrMz853viKpijDHGFCYs2AEYY4wp/yxZGGOMKZIlC2OMMUWyZGGMMaZIliyMMcYUyZKFMcaYIlmy\nKEdEZKKILC9gWBsRURGJKOu4QomItBKRZBEJL6Xy+ovIZrfMUaVRZlkSkcEisucUyyjxOhWRaiLy\ngYgkicg7pxJHeSSOH0UkOtix5BCR20TkH6VdbkgnCxHZISJp7oac83o6iPEsE5FJwZp/QSpTolHV\nXapaU1WzoVS+k/uAp90yF5xqfPnF47tDd3dAN4rIOhFJEZE9IvKOiMS6w18WkQyf7f7nU42tIL7r\ntJguBZoADVT1Mu8BIjLLK/4MEcn06v6wVIIPvEuBvaq6oSxn6m4Dm0XEIyJjfAb/B5gsIvVKc54h\nnSxcF7gbcs5rerADqmgqQ5IpRGtgfUkmPIX19iRwE3AjUB/oBCwAzvMa52Gf7T6uhPMKtNbAJlXN\n8h2gqlNz4gceBN7yWp5zyzzSkpkKzA3CfFcBU4B1vgNUNQX4DBhXqnNU1ZB9ATuAswsY9izwrlf3\nv90VKMBgYA/wf8Aht5xxXuNWBR4FdgEHgVlANa/hFwGrgWPAVmAk8ACQDaQDyTj/RgG6AJ8AR4CN\nwOVe5TQAFrrl/ADcDywvYHnaAApMcOM6BNzlNTwMuNON5zDwNlDfHbbLnTbZfZ3h9r8W+AU4CnwE\ntPYqT4FpwGZgu9vvTOBHIMl9P9Ptf4ZX2cnuOtgBnAak4vyrzCm3F5AARALhwGPusmwHprvzjXDH\nbeaunyPAFmCyVzl9gRXuujsIPO6zniLy+06AZ4DHfNbtQuCWfNb5VsADpLnTVy0ipnuBecBrblyT\n8ilzmW9/3O3R/dzRjblvIdv9y8A//fyN5Jbtdnd1Y0jESYIXeg2r5n4fO93veLnbL3edFjCPfMsE\n/gFkAJnu+vtjIXHeC7yWT/9LgA1u2Z8CHb2GHQBuB351v485QFV3WCPgQ3dbOwK8DzT1Wie+2+uv\n7rD+wPfu/PYBMwpZ7urusjX06vcQ8DrwX+A4sAboEcB94ApgTD79/wh8WKrzCtRClMWLwpNFdWAT\nMBE4C2eH1MJrY8kCHsfZAQwCUoDO7vAZODuE+kAt4APgX+6wvu4PaTjODro50MUdtgyvHQFQA9gN\nXIOz8+rpxhHtDn8TZ6deA+gG7KXoZPEczg84DjgBdHWH3wR8B7Rwl2k28F+faSO8yrsIZ2fX1Y3t\nbuAbr+GKk+Tqu/Orj5NUrnLHH+t2N/CJMxL4wmt9LQGu9xo+A3jK/TwVZ0fQAqiHszPwThZf4hxS\nRwE9cH74Q91h3wJXuZ9rAqfnt6z5fCd9cXYCYW53Q5yE1sSfbayImO7F2XmMcreNavmUlycer+1x\nj9c62VnEdv8yJUgW7nezBedPUhVgKM4OLWe7f8aNrzlOIj/T3ZZO2n58vu/CyryXfJJAPuWcNB4Q\n65Y12C37b+72kvPdHgB+wkngjXD+wNztDmuCs41XA+rgJIs385lvVXdbusdr++jjLn97d9mmFhBz\nb+CwT7+H3O1puFvGDGBZIcu9EScx5fd63I/1VlCyOBPY58824u8r6Dv8Uwre+SEn+6xg7396/XD+\nVewExvr8gLKAGl793nY3RsFJHO29hp3B7/+uZwMzCohnGXl3TFcAX/mMMxu4x92QMnETjTvsQYpO\nFi28+v2Qs6HgHCEM8xrW1C0/gvyTxYd4/dPD2bml4h5duOMP9Rp+FfCDT0zfAhN9+j0LLOL3nfEV\nwNfu53CcH3hft/tz4Dqvac/m96OCljj/sGt5Df8X8LL7+Uucf64NC1hP+SYLr3U13P08HVhSxDZ2\ntvu5qJjuBb4sYpvNL57B/L5Dvwv4rogyXsb5N+y93b9SwLjeZZ/lrv8wr+H/deMOwzmCiitk28sv\nWRRYptc6KWmyeAB41as7HCc55/wxOOC9/QEXA+sLKP90YH8+/V8E3gWkgOnuxP3Tlc+wYcAOn34P\nAYu8unsBiUUtf0lfFJwsYoHU0pxXRbhmMUpV63q9nssZoKrfA9twEsDbPtMdVefcXo6d/P4PpTqw\nUkQSRSQR+J/bH5wdxlY/Y2sN9Mspxy1rHM7pmUY4O8XdPjEU5YDX51Scf9U585rvNZ9fcHZsTQqJ\n7Umv8Y/grKfmXuN4x9Ysn/h2eo8vItfh7JyuVFWP2/t9IFpE2uL820pS1R+8yvSeh+/8jqjq8QLm\n90ecc/m/urVRzi9gOfPzCjDe/Twe/885FxUT5F2G/GTh/Bv3FomT2ME5hdjUj1ge9dnuJ/gxTTNg\nt9d3A7/H3xDnaMnfbdufMk9Vnm1OnQvseyl4fef8hhGRWiLyoojsEpFjwMc4y5hLRG4C4oEJ6u5h\nRSRaRD4UkYPudH/3nc7LUZwzD74K+o2WpVo4fyJKTUVIFgUSkWk4h5n7cM5teqsnIjW8ulu54x3C\n+YcV4/VDrKPORThwNs72BcxSfbp3A1/4/Khrqur1OP+QsnCSj3cMJbUbONdnXlGqujefuHLGv85n\n/Gqq+k0By7MPJ8F4a4Xz40VEzsK55nKRqh7LLUA1HSdRj8c5OvHeMe/HOQWVw3td7APqi4j3jzF3\nfqq6WVXHAo1xrkfN8/k+81uGHK8BF4lIHM5pOH9rORUaUyHz87YL55+6t7b8vlP8DGghIvF+xlQc\n+4CWIuL9u8+J/xDO0UpB23ZJyjxVebY5t+puc5+yfX8/+9zPd+JsW31UtTZwDs6foZyyhgF/xdle\nk73KeA7n4nF7d7r7vKfz8StQS0QaFH/RcuPY6lOrzfv1REnLxdmuS7WGXIVNFiLSCfgnv++kbheR\nHj6j/UNEqrg7uvOBd9x/SM8BM0SksVtWcxEZ4U7zAnCNiAwTkTB3WBd32EGgnVf5i4BOInKViES6\nrz4i0tX9l/QecK+IVHfrafvz77Ags4AHRKS1G3MjEbnIHZaAc6G2nc/4fxWRGHf8OiKSp2qjjyXu\nslwpIhEicgUQDSwSkZY4CeFqVd2Uz7Sv4lw7upC8yeJt4CZ3HdYF7sgZoKq7gW+Af4lIlIh0xzma\neM2Nd7yINHK/r5x/UN7/bnP4fieo6h6c89tzcSpBpBWy3N7TFRqTn97C2X76ulVkOwG34Fy/QlU3\n41wT+a9bpbaKO68xInJnMeaTn+9x/une7m6Lg4ELcM7le3BOyTwuIs1EJFxEzhCRqiUt8xRjBWdd\njRaRgSISiZMADuOceslxo4g0FZGG7vC33P613LgS3WF350zgHuW+jnMEvN1nnrVwjn6T3d/G5IKC\nU9VUnNOKg0q6gKraXvPWavN+3VzQdDnbBU4ii3S3Ee+kNgjnVHPpCdS5tLJ44ZxPzqmpkvOaj3N6\n5wfgTq9xrwfW4hxpDMapDXUXzj+qXbgXS91xo3CuH2zDqdXyC3Cj1/DROLUcjuNcABvh9j8D56L6\nUWCm268zsBhnh30Y5zx9D3dYI5yEUpzaUN7XHZbhnv/GSfy34lwwO45zOuFBr3Hvc2NI5Pdzvle5\n6+QYzpHGi17jK9DBJ4YBwEqcC/wrgQFu/4k4O2rv72G9z7SbcY6yvPtF4FwAPIxTG+oWnNMx4g5v\n4a6fI+7yTPWa9jXgt5x54ZyOPGk95feduP3Hu+MN8WMb877AXVhM9+Lf+flr3ZiPudvPneQ95y84\nFRbW4+zw9uLsBGPc4S/j1DLyXt+HCpjXYPLWhorBqYCQhHOxeLTXsGrAE+78knCuC/lTG6qwMv1d\nJ/mOB1yO8/tLxPntdPYa5l0b6ijwPBDlDmuFU5sr2R3+JyDLHTYV5xSt9/pb6Q4b5m4vyTi/rweB\nTwuJ+xJgvlf3Q8DzXt1dcuZbyvu+79zvxPuV87uugXsUXJrzzPlRViruv5/XVLVFUeOa0iEinwNv\nqOrzhYxzLjBLVX1PdwUinoE4Cae1VsYfQQUgIgeAS1U131YPyigGwTm6mqhlfGNeQUTkNpxKGH8v\nzXIr881WpoyISB+cWiEX+fSvBgzBufjYBKeW2PwyiCcS55/785YozKlwt5++wY7Dm6o+EohyK+w1\nC1M+iMgrOPdP3Kx5axGBc7rlHzinEH7COd1Qqv+G8omnK84pjaY4p1yMMX6olKehjDHGFI8dWRhj\njCmSJQtjjDFFCrkL3A0bNtQ2bdoEOwxjjAkpK1euPKSqjYoeM38hlyzatGnDihUrih7RGGNMLhHx\npzmhAtlpKGOMMUWyZGGMMaZIliyMMcYUKeSuWeQnMzOTPXv2kJ6eHuxQjAk5UVFRtGjRgshI35bT\njfldhUgWe/bsoVatWrRp04a8DS8aYwqjqhw+fJg9e/bQtm3bYIdjyrEKcRoqPT2dBg0aWKIwpphE\nhAYNGthRuSlShUgWgCUKY0rIfjvGHxUmWYSSl19+menTp5do2okTJzJv3rwiy9+3b1+h4/jasWMH\n3bp1y7f/G2+8UeT0+/bt49JLLwVgxYoV3HjjjcWa/6l44oknSE1Nze2uWbN4T7FctmwZ33zzTdEj\nngLfecyaNYtXX321VMp+8MEHS6UcYwpTOZPF669DmzYQFua8v/56sCMqVSVJFgXxN1k0a9YsN4nF\nx8czc+bMUpm/P3yTRXEFI1lMnTqVq6++ulTKLm6yUFU8nvweKmhMwSpfsnj9dZgyBXbuBFXnfcqU\nU04Yr776Kt27dycuLo6rrroKgA8++IB+/frRs2dPzj77bA4ePHjSdAcPHmT06NHExcURFxfHN998\nc9K//EcffZR77733pGnvu+8++vTpQ7du3ZgyZQqqyrx581ixYgXjxo2jR48epKWlsXLlSgYNGkTv\n3r0ZMWIE+/fvB2DlypW5833mmWfyXa4777yTr776ih49ejBjxgx27NjBWWedRa9evejVq1fuDtA7\n5mXLlnH++ecDcO+99zJhwgTOOussWrduzXvvvcftt99ObGwsI0eOJDMzs8BlAZg5cybR0dF0796d\nMWPGnBTfzJkz2bdvH0OGDGHIkCG5/e+66y7i4uI4/fTTc9d7QkICl1xyCX369KFPnz58/fXX7Nix\ng1mzZjFjxgx69OjBV1995df3lrP+8luvvjHnN497772XRx99FIDBgwdzyy23EB8fT9euXfnxxx+5\n+OKL6dixI3ffnfs0UEaNGkXv3r2JiYlhzpw5ud9PWloaPXr0YNy4cQA8/vjjdOvWjW7duvHEE0/k\nfj+dO3fm6quvplu3buzevTvfZTKmQKX9uL9Av3r37q2+NmzYcFK/ArVureqkibyv1q39L8PHunXr\ntGPHjpqQkKCqqocPH1ZV1SNHjqjH41FV1eeee05vvfVWVVV96aWXdNq0aaqqevnll+uMGTNUVTUr\nK0sTExN1+/btGhMTk1v+I488ovfcc4+qqk6YMEHfeeedPPNRVR0/frwuXLhQVVUHDRqkP/74o6qq\nZmRk6BlnnKG//fabqqq++eabes0116iqamxsrH7xxReqqvqXv/wlzzxzLF26VM8777zc7pSUFE1L\nS1NV1U2bNmnO9+Eds/c099xzj/bv318zMjJ09erVWq1aNV2yZImqqo4aNUrnz59f6LI0bdpU09PT\nVVX16NGj+a7/1q1b5657VVUgd/rbbrtN77//flVVHTt2rH711Veqqrpz507t0qVLboyPPPJI7vQF\nfW/eCluv+cXsOw/v7kGDBuntt9+uqqpPPPGENm3aVPft26fp6enavHlzPXToUJ51lJqaqjExMbn9\na9SokVvuihUrtFu3bpqcnKzHjx/X6OhoXbVqlW7fvl1FRL/99tt812GxfkMmJAEr9BT2vRWi6myx\n7NpVvP5++Pzzz7nsssto2LAhAPXr1wecKr1XXHEF+/fvJyMjI9+qiZ9//nnuuevw8HDq1KnD0aNH\n/Zrv0qVLefjhh0lNTeXIkSPExMRwwQUX5Bln48aNrFu3juHDhwOQnZ1N06ZNSUxMJDExkYEDBwJw\n1VVX8eGHRT/fPTMzk+nTp7N69WrCw8PZtGlTkdOce+65REZGEhsbS3Z2NiNHjgQgNjaWHTt2FLos\n3bt3Z9y4cYwaNYpRo0b5tV6qVKmSe2TTu3dvPvnkEwA+/fRTNmz4/cmXx44dIzk5+aTp/fneClqv\nQIlivvDCC3PXSUxMTG5Z7dq1Y/fu3TRo0ICZM2cyf77zIMHdu3ezefNmGjRokKec5cuXM3r0aGrU\nqAHAxRdfzFdffcWFF15I69atOf300/2KxxhfATsNJSIvishvIrKugOEiIjNFZIuIrBGRXoGKJY9W\nrYrX/xTccMMNTJ8+nbVr1zJ79my/qydGRETkOaec33Tp6en86U9/Yt68eaxdu5bJkyfnO56qEhMT\nw+rVq1m9ejVr167l448/LvEyzZgxgyZNmvDzzz+zYsUKMjIyipymatWqAISFhREZGZlb+yYsLIys\nrKxCl2Xx4sVMmzaNVatW0adPH7KyshgxYgQ9evRg0qRJ+c7Pex7h4eFkZWUB4PF4+O6773LXxd69\ne/O9GF7Q9+Y938LWa34xF2cd5Xz2XkfLli3j008/5dtvv+Xnn3+mZ8+exa7umpNAjCmJQF6zeBkY\nWcjwc4GO7msK8GwAY/ndAw9A9ep5+1Wv7vQvoaFDh/LOO+9w+PBhAI4cOQJAUlISzZs3B+CVV17J\nd9phw4bx7LPOomdnZ5OUlESTJk347bffOHz4MCdOnGDRokUnTZezo2jYsCHJycl5akjVqlWL48ed\nJ5h27tyZhIQEvv32W8A5Mli/fj1169albt26LF/uPOv+9QKu2XiXlbNMTZs2JSwsjLlz55Kdne3n\nWipYQcvi8XjYvXs3Q4YM4d///jdJSUkkJyfz0UcfsXr1ap5//vl8YyzIOeecw1NPPZXbvXr16gKX\nMb/vzXu+Ba3XgmL2N8aCJCUlUa9ePapXr86vv/7Kd999lzssMjIy99rPWWedxYIFC0hNTSUlJYX5\n8+dz1llnlXi+xuQIWLJQ1S+BI4WMchHwqns67Tugrog0DVQ8ucaNgzlzoHVrEHHe58xx+pdQTEwM\nd911F4MGDSIuLo5bb70VcC7uXnbZZfTu3Tv3FJWvJ598kqVLlxIbG0vv3r3ZsGEDkZGR/P3vf6dv\n374MHz6cLl26nDRd3bp1mTx5Mt26dWPEiBH06dMnd9jEiROZOnUqPXr0IDs7m3nz5nHHHXcQFxdH\njx49ci9Kv/TSS0ybNo0ePXrkXlD21b17d8LDw4mLi2PGjBn86U9/4pVXXiEuLo5ff/01z7/VktbX\nL2hZsrOzGT9+PLGxsfTs2ZMbb7yRunXrnjT9lClTGDlyZJ4L3PmZOXMmK1asoHv37kRHRzNr1iwA\nLrjgAubPn5/n4nNR31uVKlXyXa8Fxew7j+IaOXIkWVlZdO3alTvvvDPP6aQpU6bknvrq1asXEydO\npG/fvvTr149JkybRs2fPYs/PGF8BfQa3iLQBFqnqSRX4RWQR8JCqLne7PwPuUNVCH1YRHx+vvs+z\n+OWXX+jatWtphW1KYOXKldx666188cUXwQ7FlID9hkpXtkdJTs8iJSOLtMxs0jOzSc/0uO/Zbj8P\naZnZnMjMJi0jm8xsD5keJTPLQ5ZHycz2kJWtZHrc92wPmdlKVp5uD9keJVuVbI9z2jmn25P7Dh5V\nvvu/s1eqanxJlykkLnCLyBScU1W0CsC1BXNqVqxYwZVXXslDDz0U7FCMKVVZ2R6OpGZwONl9pZzI\nfT+Sksmx9EyOp2dxPD2T5PSs3M8pGSU7PRsZLkSGhxER5r6HCxFhYVSJcPpFhIcRGS65n6tXiSA8\nTAgPE8JECA+DMBHCwoRwydv/u6JnX6hgJou9QEuv7hZuv5Oo6hxgDjhHFoEPzRRHfHy8X7WijClv\nUk5ksf1QCnuOprI3MZ19iWm5r72J6RxKPpHvdOFhQr3qkdSOiqRWVAS1oiI5rXZU7uec9xpVwqlW\nJZyoSPcVEfZ7d0Q4UVXCcj9HhktAm1555BSnD2ayWAhMF5E3gX5AkqruD2I8xpgKKvlEFhv2HeOX\n/cfYmpDM1oRktiWksD8pb42yapHhNKsbRbO61ejatDZNakfRsGYVGtSsSoMaznvDmlWoHRVJWFjl\nalMrYMlCRP4LDAYaisge4B4gEkBVZwFLgD8AW4BU4JpAxWKMqTwysz2s3ZvEih1HWLv3GOv3JrH9\ncAo5l2drRUXQrlFNzmjXgPaNa9KuYQ1a1q9Oi3rVqFMt0hpWLEDAkoWqji1iuALTAjV/Y0zlkO1R\nftp1lK+3HOaHHYdZtTORtEznmkHzutXo1rw2o3s2p1vzOkQ3q03jWlUtIZRASFzgNsYYb0lpmXyx\nKYHPfznIsk0JJKZmIgJdT6vNFX1a0q9tffq0rU/DmlWLLsz4pfI1JFiB+NsirK+CmjkfPHgwvtWS\nAzH/8mThwoW5tbgWLFiQpzkQf/z666/06NGDnj17snXr1lKLy7cl3T/84Q8kJiaWWvmhKC0jm0Vr\n9jH51RXE//MTbvzvT3y5+RBDuzTm6St7svpv57DkprO498IYzo1taomilNmRRQjL2VlfeeWVITP/\nrKwsIiLKz2Z34YUX5rbLtGDBAs4//3yio6P9nn7BggVceumleVqHLQ1PPPEE48ePp7rb2sCSJUtK\ntfxQsnZPEnO/28HiNftJycimSe2qTDijDefGNqVHy7qEV7ILzUFzKq0QBuN1yq3OBsjcuXO1T58+\nGhcXp1OmTNGsrCzdsWOHdujQQRMSEjQ7O1sHDBigH330kW7fvl07d+6sV155pXbp0kUvueQSTUlJ\nUVWn1dCBAwdqr1699JxzztF9+/apqurmzZt12LBh2r17d+3Zs6du2bJF+/Xrp7Vr19a4uDh9/PHH\nNSsrS//yl79ofHy8xsbG6qxZs1RV1ePx6LRp07RTp046bNgwPffcc3NbrvWW0/ppnz59tGPHjvrl\nl1+qqhZYrr/zX7p0qQ4YMEAvuOAC7dixo6qqPvbYYxoTE6MxMTG5re4+++yzGhcXp3FxcdqmTRsd\nPHiwvvDCC3rTTTflxjhnzhy9+eabVVX1vvvu006dOmn//v11zJgxua24/vTTT9qvXz+NjY3VUaNG\n6ZEjR1RV9cknn9SuXbtqbGysXnHFFar6ewvAX3/9tdarV0/btGmjcXFxumXLFu3Zs2fufDdt2pSn\nW1V18eLF2qRJE23WrJkOHjy40NaCC1u3f/7znzUmJkZjY2N15syZ+uSTT2pkZKR269ZNBw8erKp5\nW9bNb91t375du3TpopMmTdLo6GgdPny4pqamFrHV/q48/Ia8pWdm6bwVu/XCp5dr6zsWade/fai3\nvbNav96SoFnZnmCHF5I4xVZng77zL+6rqGRx78J1evmsb0r1de/CdYV+CRs2bNDzzz9fMzIyVFX1\n+uuv11deeUVVnSauL730Un344Yd1ypQpqur8sAFdvny5qqpec801+sgjjxTa7HXfvn31vffeU1XV\ntLQ0TUlJOan58NmzZ+c2x52enq69e/fWbdu26bvvvqtnn322ZmVl6d69e7VOnToFJouc5rgXL16s\nw4YNK7Rcf+e/dOlSrV69um7btk1VC25GO0dGRoYOGDBAFy5cqMePH9d27drlrtszzjhD16xZoz/8\n8IPGxcVpWlqaHjt2TDt06JCbLGJjY3XZsmWqqvq3v/0tN9nk13S4d3Px3s2/q6oOHjxYf/rpJ1VV\n/etf/6ozZ848aZ15NzVeVLLIb93+5z//0UsuuUQzMzNV9fdmyH2bXc/pLqwJ8vDw8Nx4L7vsMp07\nd+5J8RakvCSLtIwsfWn5Nj39wU+19R2LdOijS/Wl5ds0KS0j2KGFvFNNFuXnfEAI++yzz1i5cmVu\nm0ZpaWk0btwYgEmTJvHOO+8wa9as3IbrAFq2bEn//v0BGD9+PDNnzmTkyJH5Nnt9/Phx9u7dy+jR\nowGIiorKN46PP/6YNWvW5F6PSEpKYvPmzXz55ZeMHTuW8PBwmjVrxtChQwtclosvvhhwmvbOaT68\noHKrVKni1/yrVKlC3759c5v6LqgZ7Zw2jG666SaGDh2a29z60KFDWbRoEV27diUzM5PY2FieeOIJ\nLrroIqKiooiKisodNykpicTERAYNGgTAhAkTuOyyy4DiNx0+adIkXnrpJR5//HHeeustfvjhhyKn\nKUx+6/bTTz9l6tSpuafmcpq3L0hhTZC3bduWHj16nDSPUHAiK5u53+5k1hfbOJR8gr5t6vPQJd0Z\n2LGh1VwqJypcsrjngpgyn6eqMmHCBP71r3+dNCw1NZU9e/YA5LY+Cic3uiciuc1e57RkmsPf1kpV\nlaeeeooRI0bk6V+c8905zWN7N+1dULnLli3za/7Lli3zu3nsl19+mZ07d/L000/n9ps0aRIPPvgg\nXbp04ZprSn47zuLFi/nyyy/54IMPeOCBB1i7dm2h419yySX84x//YOjQofTu3fukZ0f4Kqpp+fzW\nbWnybto8PDyctLS0Up9HaVNVlqw9wL//9yu7jqTSv0MDnhnak37tCl/XpuxZbahSMGzYMObNm8dv\nv/0GOE2U79y5E4A77riDcePGcd999zF58uTcaXbt2pWbFN544w0GDBhQYLPXtWrVokWLFixYsACA\nEydOkJqaelKz1yNGjODZZ5/Nba5606ZNpKSkMHDgQN566y2ys7PZv38/S5cuLdbyFVSuv/P3VVAz\n2itXruTRRx/ltddeIyzs902zX79+7N69mzfeeIOxY53bd/r3788HH3xAeno6ycnJuc2416lTh3r1\n6uW27Dp37lwGDRpUYNPh3nyXJyoqihEjRnD99df7laT8aVre1/Dhw5k9e3Zu8shp3r6gJs0rUhPk\nGw8c57IZ/LR+AAAgAElEQVRZ3zLtjVVUrxLOq9f25fVJp1uiKKcq3JFFMERHR/PPf/6Tc845B4/H\nQ2RkJM888ww7duzgxx9/5OuvvyY8PJx3332Xl156iSFDhtC5c2eeeeYZrr32WqKjo7n++utzm72+\n8cYbSUpKIisri5tvvpmYmBjmzp3Lddddx9///nciIyN555138jQfPnHiRG666SZ27NhBr169UFUa\nNWrEggULGD16NJ9//jnR0dG0atWKM844o1jLN2nSpHzL9Xf+vryb0c4pv2fPnlxzzTUcOXIkt6nx\n+Pj43GdWXH755axevZp69eoB0KdPHy688EK6d+9OkyZNiI2NpU6dOoDzDIqpU6eSmppKu3bteOml\nl3KbDk9KSkJV823ufMyYMUyePJmZM2cyb9482rdvz7hx45g/fz7nnHNOkevJu2n55s2b59u0fH7r\ndtOmTXTv3p3IyEgmT57M9OnTc5tdb9asWZ7kXtC6C6VTTumZ2TyzdAvPLttK7WqRPHRxLJfFt7Ra\nTeVcQJsoD4SK0ET5jh07OP/881m3Lt+HCJp8nH/++dxyyy0MGzYst19ycjI1a9YkNTWVgQMHMmfO\nHHr1Kt0HLj766KMkJSVx//33l2q55U1Z/YZ+3p3ILW+vZltCChf3bM7d50dTv0aVoic0p0xEKn4T\n5abySkxMpG/fvsTFxeVJFOA89GfDhg2kp6czYcKEUk8Uo0ePZuvWrXz++eelWm5l5PEoc77axqMf\nbaRxraq8em1fBnZqFOywTDFYsgiCNm3a2FGFn+rWrVtg8+eBvnt8/vz5AS2/skg4foJb3lrN8i2H\nOLfbaTx0cXfqVI8MdlimmCxZGGMCZt3eJKa8uoLDKRn86+JYxvRpaVVhQ1SFSRaqahuhMSUQqOuW\nH/y8j9vm/Uz96lV49/oz6da8TkDmY8pGhag6GxUVxeHDhwO20RtTUakqhw8fLvBGz5KW+fTnm7nh\nvz/RrVkd3p8+wBJFBVAhjixatGjBnj17SEhICHYoxoScqKgoWrRoUSpleTzKfYs28PI3O7i4Z3Me\nuqQ7VSIqxH/SSq9CJIvIyMjcpiSMMcGRkeXhtnk/8/7qfUwa0Jb/+0PXSvfo0YqsQiQLY0xwZWZ7\nuOG/q/ho/UHuGNmFqYPa2TXECsaShTHmlGRle7j5zdV8tP4g91wQzTX97Si/IrKTicaYEsv2KLe+\n/TOL1+7n7vO6WqKowCxZGGNKRFX52/vrWPjzPu4Y2YVJZ7ULdkgmgCxZGGNK5D/LtvLG97uYOqg9\n1w9uH+xwTIBZsjDGFNu7K/fwyEcbGdWjGbeP6BzscEwZsGRhjCmWb7Yc4o5319C/QwMevjTOqsdW\nEpYsjDF+230klWlvrKJNwxo8O7633XBXidg3bYzxS1pGNtfNXUmWR3nu6nhqR1nLsZVJkclCRFaK\nyDQRqVcWARljyh9V5c731vDLgWM8OaYHbRv690x1U3H4c2RxBdAM+FFE3hSREWK3ZhpTqbz67U7e\nX72PW8/uxNAuTYIdjgmCIpOFqm5R1buATsAbwIvAThH5h4jUL2xaERkpIhtFZIuI3JnP8Doi8oGI\n/Cwi60XkmpIuiDEmMDbsO8YDi39haJfGTBvSIdjhmCDx65qFiHQHHgMeAd4FLgOOAQU+b1JEwoFn\ngHOBaGCsiET7jDYN2KCqccBg4DERsQfyGlNOpGZkMf2/q6hbPZJHLu1uNZ8qsSLbhhKRlUAi8AJw\np6qecAd9LyL9C5m0L7BFVbe55bwJXARs8BpHgVruaa2awBEgq9hLYYwJiHsXrmf7oRRen9SPBjWr\nBjscE0T+NCR4Wc4OP4eItFXV7ap6cSHTNQd2e3XvAfr5jPM0sBDYB9QCrlBVj29BIjIFmALQqlUr\nP0I2xpyqRWv28faKPUwf0oEz2zcMdjgmyPw5DTXPz34lMQJYjXMBvQfwtIjU9h1JVeeoaryqxjdq\n1KiUZm2MKUjC8RP8bcE64lrW5aazOwY7HFMOFHhkISJdgBigjoh4H0HUBvx5BuNeoKVXdwu3n7dr\ngIfUeR7qFhHZDnQBfvCjfGNMAKgqf1uwjpSMbB67rDuR4XY7lin8NFRn4HygLnCBV//jwGQ/yv4R\n6CgibXGSxBjgSp9xdgHDgK9EpIk7z20YY4LmgzX7+d/6A9x5bhc6NK4V7HBMOVFgslDV94H3ReQM\nVf22uAWrapaITAc+AsKBF1V1vYhMdYfPAu4HXhaRtYAAd6jqoZIsiDHm1CUcP8E97zunnyYNsGdT\nmN8VdhrqdlV9GLhSRMb6DlfVG4sqXFWXAEt8+s3y+rwPOKdYERtjAubehetzTz9F2Okn46Ww01C/\nuO8ryiIQY0xwLdv4G4vX7ufPwzvZ6SdzksJOQ33gvr+S009EwoCaqnqsDGIzxpSR9Mxs7lm4nnaN\najBlkD3xzpzMn4YE3xCR2iJSA1gHbBCR2wIfmjGmrPxn2VZ2Hk7lnxd1o2pEeLDDMeWQPyclo90j\niVHAh0Bb4KqARmWMKTPbEpKZtWwrF/Voxpkd7OY7kz9/kkWkiETiJIuFqpqJ00yHMSbEqSr3LFxP\n1cgw7jqva7DDMeWYP8liNrADqAF8KSKtcRoRNMaEuE82HOSrzYf48/BONK7lz722prIqsm0oVZ0J\nzPTqtVNEhgQuJGNMWcjI8vDgkl/o0Lgm409vHexwTDnnzwXuOiLyuIiscF+P4RxlGGNC2Kvf7mDH\n4VTuOq+r3VNhiuTPFvIiThMfl7uvY8BLgQzKGBNYR1MymPnZZgZ2asSQzo2DHY4JAf40Ud5eVS/x\n6v6HiKwOVEDGmMB74tNNpGRkc7dd1DZ+8ufIIk1EBuR0uA88SgtcSMaYQNry23Fe+34XY/u2pFMT\nu1Pb+MefI4vrgVdEpA5OY39HgImBDMoYEzj//t9GqkeGc8vZnYIdigkh/tSGWg3E5TyUyJr6MCZ0\nrdp1lE82HOTPwzvZY1JNsfjzDO5bfboBkoCVbiIxxoQAVeWR/22kYc0qXGvNj5ti8ueaRTwwFeeZ\n2s2B64CRwHMicnsAYzPGlKLlWw7x7bbDTBvSgRpV/TkDbczv/NliWgC9VDUZQETuARYDA4GVwMOB\nC88YUxpUlUc+2kjzutW4sl+rYIdjQpA/RxaNgRNe3ZlAE1VN8+lvjCmn/rfuAGv2JHHz2R2tVVlT\nIv4cWbwOfC8i77vdFwBvuE2WbwhYZMaYUpGV7eHRjzfSoXFNLu7VItjhmBDlT22o+0XkQ6C/22uq\nquY8PW9cwCIzxpSKBav3sTUhhVnjexEeJsEOx4Qof2pDPQa8qKpPlkE8xphSlJXt4ZmlW4huWpsR\nMacFOxwTwvy5ZvELMEdEvheRqe7NecaYEPDBmn1sP5TCjcM65lR7N6ZEikwWqvq8qvYHrgbaAGvc\nR61aM+XGlGPZHuWpz7fQ5bRanBPdJNjhmBDnV7vEIhIOdHFfh4CfgVtF5M0AxmaMOQWL1uxjW4Jz\nVBFm1yrMKfLnmsUMnBpQnwEPquoP7qB/i8jGQAZnjCkZj3tU0alJTUbatQpTCvypOrsGuFtVU/IZ\n1reU4zHGlIIP1x1gy2/JPDW2px1VmFLhz2mo8cAg7x4iMgdAVZMCEZQxpuQ8HmXmZ5tp36gGf4ht\nGuxwTAXhT7JoC9zhNvORIz5A8RhjTtHHGw6w8eBxbhja0e6rMKXGn2SRCAwDmojIB8WpOisiI0Vk\no4hsEZE7CxhnsIisFpH1IvKFv2UbY06mqsz8bAvtGtbggrhmwQ7HVCD+JAtR1SxV/RPwLrAcp72o\nwidyalA9A5wLRANjRSTaZ5y6wH+AC1U1BrismPEbY7x8tfkQG/YfY+qg9nZUYUqVP8liVs4HVX0Z\n5yl5H/sxXV9gi6puU9UM4E3gIp9xrgTeU9Vdbvm/+VGuMaYAs7/cSpPaVbmopx1VmNLlz015s0Uk\nTkSmi8h0IEtVr/Wj7ObAbq/uPW4/b52AeiKyTERWisjV+RUkIlNEZIWIrEhISPBj1sZUPmv3JPH1\nlsNc27+ttSxrSl2RyUJEbsRpebax+3pNRG4opflHAL2B84ARwN9E5KQHA6vqHFWNV9X4Ro0aldKs\njalYZn25lVpVIxhrz6swAeDPfRaTgH4591mIyL+Bb4GniphuL9DSq7uF28/bHuCwW3aKiHwJxAGb\n/IjLGOPaeTiFD9fuZ/LAdtSOigx2OKYC8usCN5Dt1Z3t9ivKj0BHEWkrIlWAMcBCn3HeBwaISISI\nVAf64TRcaIwphue/2k5EWBjX9rdna5vA8OfI4iWchx/Nd7tHAS8UNZGqZrnXOD4CwnGaOV8vIlPd\n4bNU9RcR+R/OXeIe4HlVXVeSBTGmsjqcfIK3V+xmdM/mNKkdFexwTAXlz8OPHheRZcAAt9c1qvqT\nP4Wr6hJgiU+/WT7djwCP+BWtMeYkr3yzgxNZHiYPbBfsUEwFVmiycO+VWK+qXYBVZROSMcZfKSey\neOXbnQyPbkKHxjWDHY6pwAq9ZqGq2cBGEbHqFcaUQ2+v2E1SWiZTB7UPdiimgvPnmkU9YL2I/ADk\ntjyrqhcGLCpjTJEysz08/9V2+rSpR+/W9YIdjqng/EkWfwt4FMaYYlu8Zj97E9P4x4UxwQ7FVAL+\nXOD+QkRaAx1V9VO3iqvdHmpMEKkqs77YSsfGNRnapcim2ow5Zf7cwT0ZmAfMdns1BxYEMihjTOG+\n3HyIXw8cZ8rAdvZwI1Mm/LkpbxrQHzgGoKqb8aPVWWNM4Mz+wm0wsIdvc2vGBIY/yeKE22osACIS\nAWjgQjLGFGbNnkS+2XqYPw5oS5UIf37Cxpw6f7a0L0Tk/4BqIjIceAf4ILBhGWMKMvuLbU6DgX2t\nRrspO/4kizuBBGAtcB2wRFXvCmhUxph87TiUwofr9jPu9NbUsgYDTRnyp+rsDar6JPBcTg8Rucnt\nZ4wpQ88v3+Y2GNgm2KGYSsafI4sJ+fSbWMpxGGOKcCj5BO+s2MPFvZrT2BoMNGWswCMLERmL89jT\ntiLi3bR4LeBIoAMzxuT1yjc7yMi2BgNNcBR2GuobYD/QEHjMq/9xnCbFjTFlJOVEFq9+u5PhXZvQ\nvpE1GGjKXoHJQlV3AjuBM3zu4K4GVMNJGsaYMvDWj26DgYOtwUATHCW5g7sFdge3MWUmM9vDC8u3\n07dNfXq1sgYDTXDYHdzGlHOL1uxjb2Ia1w2yaxUmeOwObmPKMVVl9hfb6Ni4JkM62380Ezx2B7cx\n5diyTQnWYKApF0p0BzdwdyCDMsY4Zi3bStM6UdZgoAk6f55n4cG5e/u5osY1xpSen3Yd5fvtR7j7\nvK7WYKAJuiKThYj0B+4FWrvjC6CqalfbjAmg2V9so3ZUBGOswUBTDvjTNtQLwC3ASiA7sOEYYwC2\nJiTz0YYDTBvcgZpV/fmZGhNY/myFSar6YcAjMcbkeu7LbVQJD2OiNRhoyonC2obq5X5cKiKPAO8B\nJ3KGq+qqAMdmTKX027F03lu1l8v7tKBhzarBDscYoPAji8d8uuO9PiswtPTDMca8+PUOsjweJp9l\nlwVN+VFY21BDyjIQYwwcS8/k9e92cm5sU1o3qBHscIzJFdD6eCIyUkQ2isgWEbmzkPH6iEiWiFwa\nyHiMKe/e+H4Xx09kcf0gazDQlC8BSxYiEg48A5wLRANjRSS6gPH+DXwcqFiMCQUnsrJ5cfl2BnRo\nSLfmdYIdjjF5BPLIoi+wRVW3uW1LvQlclM94NwDvAr8FMBZjyr0FP+3lt+MnmGpHFaYcKqw21MWF\nTaiq7xVRdnNgt1f3HqCfzzyaA6OBIUCfQmKZAkwBaNXKblAyFU+2R5n1xTa6Na9N/w4Ngh2OMScp\nrDbUBe57Y+BM4HO3ewjOU/SKShb+eAK4Q1U9IgU3kqaqc4A5APHx8dbiralwFq3Zx/ZDKcwa35vC\nfgvGBEthtaGuARCRj4FoVd3vdjcFXvaj7L1AS6/uFm4/b/HAm+6PoyHwBxHJUlV7uJKpNDwe5Zml\nW+jUpCbnRDcJdjjG5MufO7hb5iQK10HAn3NBPwIdRaQtTpIYA1zpPYKqts35LCIvA4ssUZjK5uMN\nB9h0MJknx/SwZshNueVPsvhMRD4C/ut2XwF8WtREqpolItOBj4Bw4EVVXS8iU93hs0oYszEVhqry\n1OdbaNuwBud3bxbscIwpkD9NlE8XkdHAQLfXHFWd70/hqroE5/kX3v3yTRKqOtGfMo2pSJZu/I31\n+47xyKXdCbejClOO+duc5SrguKp+KiLVRaSWqh4PZGDGVHSqyszPttCiXjVG9bSHG5nyrcj7LERk\nMjAPmO32ag7YdQVjTtHXWw6zenci1w9uT2S4PdzIlG/+bKHTgP7AMQBV3YxTndYYcwqe+nwzp9WO\n4tLeLYIdijFF8idZnHDvwAZARCJwWp01xpTQN1sO8f32I1w3qB1VI8KDHY4xRfInWXwhIv8HVBOR\n4cA7wAeBDcuYiktVeeyTTZxWO4qx9shUEyL8SRZ3AgnAWuA6YImq3hXQqIypwL7YlMDKnUeZPrQD\nUZF2VGFCgz+1oW5Q1SeB53J6iMhNbj9jTDGoKo99vIkW9apxeXzLoicwppzw58hiQj79JpZyHMZU\nCh9vOMjavUncNKwjVSKsBpQJHYW1OjsWp3mOtiKy0GtQLeBIoAMzpqLxeJQZn2yiXcMajLb7KkyI\nKew01DfAfpwG/ryfx30cWBPIoIypiBav3c+vB47z5JgeRNh9FSbEFNbq7E5gp4iMA/apajqAiFTD\naUF2R5lEaEwFkJntYcYnm+jUpCYXWBtQJgT58/fmbcDj1Z2NU33WGOOnN3/YxbZDKdw+oou1LGtC\nkj/JIsL7pjz3c5XAhWRMxXI8PZMnPt1Mv7b1GdbVGj8wocmfZJEgIhfmdIjIRcChwIVkTMUy+4tt\nHE7J4K7zutpT8EzI8uc+i6nA6yLyDE4zH3uAqwMalTEVxIGkdJ5fvo0L45rRvUXdYIdjTIn58zyL\nrcDpIlLT7U4OeFTGVBCPfbwRjwduG9E52KEYc0r8aaK8iYi8ALyjqskiEi0ifyyD2IwJab/sP8a8\nVXuYcGZrWtavHuxwjDkl/lyzeBnn0ag59f02ATcHKiBjKgJV5Z6F66lbLZJpQzoEOxxjTpk/yaKh\nquZWn1XVLJzqs8aYAiz8eR8/bD/CbSO6ULe6VR40oc+fZJEiIg1wn2EhIqcDSQGNypgQlnIiiweX\n/EJs8zpc0ccaCzQVgz+1oW4FFgLtReRroBFwaUCjMiaEPfX5Fg4eO8Gz43sTbjfgmQrCn9pQq0Rk\nENAZEGCjqmYGPDJjQtDWhGReWL6NS3u3oFeresEOx5hSU2SyEJEo4E/AAJxTUV+JyKyctqKMMQ5V\n5e/vryMqIpw7RnYJdjjGlCp/rlm8CsQATwFPu5/nBjIoY0LRvJV7+HrLYW4/twuNalUNdjjGlCp/\nrll0U9Vor+6lIrIhUAEZE4oSjp/gn4t/Ib51PcbZc7VNBeTPkcUqtwYUACLSD1gRuJCMCT33frCe\ntIxsHrqku7Uqayokf44segPfiMgut7sVsFFE1gKqqt0DFp0xIeCTDQdZvGY/fx7eiQ6NawY7HGMC\nwp9kMbKkhYvISOBJIBx4XlUf8hk+DrgDp5bVceB6Vf25pPMzpqwdTcngrvlr6dykFtcNah/scIwJ\nGH9OQ3VU1Z3eL2Cw1+d8iUg48AxwLhANjBWRaJ/RtgODVDUWuB+YU7LFMKbsqSp3LVjL0dQMHrs8\njioR9qhUU3H5s3X/XUSeFZEabqOCHwAX+DFdX2CLqm5zH5j0JnCR9wiq+o2qHnU7v8N5XKsxIWH+\nT3tZsvYAN5/diW7N6wQ7HGMCyp9kMQjYCqwGlgNvqKo/d3A3B3Z7de9x+xXkj8CHfpRrTNDtOZrK\nPe+vp0+beky100+mEvDnmkU9nKOErTj//FuLiKiqllYQIjIEJ1kMKGD4FGAKQKtWVi3RBFdWtodb\n3/oZBR6/vIc16WEqBX+OLL4D/qeqI4E+OE2Vf+3HdHsB71bUWrj98hCR7sDzwEWqeji/glR1jqrG\nq2p8o0aN/Ji1MYHz2Ceb+GHHEe4fFWPPqTCVhj9HFmer6i4AVU0DbhSRgX5M9yPQUUTa4iSJMcCV\n3iOISCvgPeAqVd1UrMiNCYLPfjnIs8u2MrZvK0b3tEtspvLwJ1nsFpHxQDtVvc/dwRfZLpSqZonI\ndJwHJ4UDL6rqehGZ6g6fBfwdaAD8x32QfZaqxpdwWYwJqN1HUrn17Z+JaVabey7wrdhnTMUmRV16\nEJFncR58NFRVu4pIPeBjVe1TFgH6io+P1xUr7AZyU7bSMrK5bPY37DycyqIbBtC6QY1gh2RMsYjI\nylP5M+7PkUU/Ve0lIj8BqOpREbFHf5lKw+NRbn17Nev3HeP5q+MtUZhKyZ8L3JnuDXY5T8prhPuI\nVWMqg8c/2cSH6w5w1x+6Mqxrk2CHY0xQ+JMsZgLzgcYi8gDOvRYPBjQqY8qJ+T/t4emlW7giviV/\nHNA22OEYEzT+PCnvdRFZCQzDacNplKr+EvDIjAmyZRt/47Z31nB6u/rcP6obbiUMYyolf65ZoKq/\nAr8GOBZjyo2VO48w9bWVdGpSizlXx1u7T6bSs1+AMT5+PXCMa176kdNqR/HKtX2pHRUZ7JCMCTpL\nFsZ42bDvGFc+9z3Vq0Qw94/97PGoxrgsWRjjWrsnibHPfUfViDD+O+V0a8rDGC9+XbMwpqJbteso\nE178gdpRkbxpicKYk9iRhan0Pl5/gCuf+476Narw9tQzLFEYkw87sjCV2ktfb+e+RRvo3qIuL0yI\np2FNu0ZhTH4sWZhKKSPLwwOLN/DKtzsZHt2EmWN6Uq1KeLDDMqbcsmRhKp0DSen86fWVrNqVyKQB\nbfnrH7raA4yMKYIlC1OpfL3lEDe9+ROpGdk8fWVPzu/eLNghGRMSLFmYSiE9M5t//+9XXvp6B+0b\n1eDNKafToXGtYIdlTMiwZGEqvNW7E/nz26vZmpDCxDPbcMfILnZ9wphismRhKqzE1Awe+Wgjb/yw\niya1opj7x76c1dGe4W5MSViyMBVOtkd5d+UeHvrfrySlZXLNmW25ZXhHalkbT8aUmCULU2GoKh9v\nOMhjH29k08Fk4lvX476LuhHdrHawQzMm5FmyMCHP41GWbvyNpz7fwurdibRrVIP/jOvFyJjTCLMq\nscaUCksWJmSlZ2az4Ke9PPfVNrYmpNC8bjUevqQ7F/dqTkS4tWRjTGmyZGFCzuaDx3nrx92899Ne\njqRkENOsNk+O6cF5sU0tSRgTIJYsTEhIOH6Cj9Yf4N1Ve/hpVyKR4cLZXZtw1emtOaN9A3vkqTEB\nZsnClFt7E9P4eP0BPlx7gB93HkEVOjSuyd3ndWV0z+Y0sEb/jCkzlixMuXEsPZNvtx7m6y2HWL75\nENsOpQDQuUktbhzakXNjT6Nzk1p2FGFMEFiyMEHh8ShbEpJZvSuRn3Yn8tOuo2w6eByPQvUq4fRr\nW59xp7dmSOdGtGtUM9jhGlPpWbIwAaWqHErOYPPB42w8eJyNB5z3TQeOk5KRDUDtqAh6tKrHiJjT\nOLN9A3q2qkeVCLtQbUx5Yski1Lz+Otx1F+zaBa1awQMPwLhxQQvH41GOpmaQkHyCfYlp7Dqcyu6j\naew6ksruI6nsOpJKqpsUAOpVj6TzabW4tHcLYlvUpWerurRtUMPuhzCmnAtoshCRkcCTQDjwvKo+\n5DNc3OF/AFKBiaq6KpAxhbTXX4cpUyA11eneudPphlJJGB6PkpyRRVJqJsfSM0lKy+RYWhbH0pzu\nxNRMEo6fICH5hPN+/ASHkk+Q5dE85VSvEk7LetVpWb86Z7ZvSMv61ejYuBadT6tFw5pV7JqDMSEo\nYMlCRMKBZ4DhwB7gRxFZqKobvEY7F+jovvoBz7rv5Zaqogqa8xlQd1+p6O+f9ffuk8b1GpatSrZH\nyfIo2dlKpsfjdGcrWR6P09+jZGZ7yJ45l6zTupIt4WSFh5MVFk62hJP53EKyO51FljtdZraH9Mxs\n0jKzScvwkJaZ7XRnuP3cbu9x0jOzSc3Iwme/n0d4mNCwZhUa1apKo5pV6dq0Vu7nxrWjaFI7itYN\nqtOghiUEYyqaQB5Z9AW2qOo2ABF5E7gI8E4WFwGvqqoC34lIXRFpqqr7Cyv4g5/38c/FG7x2xOB8\nIt+dc85n8hvmtUMnd5yCE0JQDbmh4GHvrj2pV3iYUD0ynKgq4URFhlEtMpxqkeFERYZTv0aV3O6q\n7nv1KuHUqRZJnWqR1M59j8jtV6NKhJ0uMqaSCmSyaA7s9urew8lHDfmN0xzIkyxEZAowBaBVq1Y0\nrRPF4E6NEQHnD6w4n3PHB/Hql/Mv9+T+zjC3CNxPeYd7l+2OW1A5FDj/37vd5cntHx4mhIcJkWFh\nhIcJEeFCRM5n3+4rLidi3x4iPNmEezxEerII92QT0fQ0Ir5eToRbVkR4GNWrhBNpdzMbY0pJSFzg\nVtU5wByA+Ph4jW9Tn/g29YMcVRDccm3eaxYA1avD/z0MdasFLy5jTIUXyL+ee4GWXt0t3H7FHcfk\nGDcO5syB1q2dw5LWrZ3uINaGMsZUDoE8svgR6CgibXESwBjgSp9xFgLT3esZ/YCkoq5XVHrjxlly\nMMaUuYAlC1XNEpHpwEc4VWdfVNX1IjLVHT4LWIJTbXYLTtXZawIVjzHGmJIL6DULVV2CkxC8+83y\n+qzAtEDGYIwx5tRZdRljjDFFsmRhjDGmSJYsjDHGFMmShTHGmCJZsjDGGFMkSxbGGGOKZMnCGGNM\nkSxZGGOMKZIlC2OMMUWyZGGMMaZIliyMMcYUyZKFMcaYIomWi+eF+k9EEoCdQEPgUAmLqQMkBWC6\ngobn17+k8Qcq9sLG8e1v6770p7N1f2oxnMp0lWXdd1bVWiWIwaGqIfkCVpzCtHMCMV1Bw/PrX9L4\nAz2E0lcAAAUeSURBVBV7ceK3dW/r3tZ9+Yi/LNZ9zquynob6IEDTFTS8pPMrzbL8mS6U4w/l2Asb\nJxTiD+XYCxsnFOIvi9iBEDwNlUNEVqhqfLDjKKlQjj+UY4fQjj+UY4fQjj+UY4dTjz+UjyzmBDuA\nUxTK8Ydy7BDa8Ydy7BDa8Ydy7HCK8YfskYUxxpiyE8pHFsYYY8qIJQtjjDFFsmRhjDGmSBU2WYhI\nDRFZISLnBzuW4hKRriIyS0Tmicj1wY6nOERklIg8JyJvicg5wY6nuESknYi8ICLzgh2LP9zt/BV3\nnY8LdjzFFWrr21sF2NaLt585lZs0AvECXgR+A9b59B8JbAS2AHf6Uc59wO3A+aEYvztNGPBaiMZe\nD3ghhNf9vLKMvaTLAVwFXOB+fitYMZ/q9xDM9V0KsZf5tl7K8fu1nwn6wuWzsAOBXt4LC4QDW4F2\nQBXgZyAaiAUW+bwaA8OBMcDEICSLU47fneZC4EPgylCL3Z3uMaBXKK57d7pgJoviLMdfgR7uOG8E\nK+aSxl8e1ncpxF7m23ppxV+c/UwE5YyqfikibXx69wW2qOo2ABF5E7hIVf8FnHSaSUQGAzVwfkxp\nIrJEVT2BjDtHacTvlrMQWCgii4E3AhdxnnmWxroX4CHgQ1VdFdiI8yqtdR9sxVkOYA/QAlhNOTmt\nXMz4N5RtdIUrTuwi8gtB2tYLUtx1X5z9TLnYuPzQHNjt1b3H7ZcvVb1LVW/GWfjnyipRFKJY8YvI\nYBGZKSKzgSWBDq4IxYoduAE4G7hURKYGMjA/FXfdNxCRWUBPEflroIMrhoKW4z3gEhF5lgA08VCK\n8o2/HK9vbwWt+/K2rRekoHVfrP1MuTuyKE2q+nKwYygJVV0GLAtyGCWiqjOBmcGOo6RU9TBQnn/4\neahqCnBNsOMoqVBb394qwLa+jGLsZ0LlyGLv/7d396xRRAEUht+DKFEihoCNKGhhlXTaBButBMEu\nYGMRtBBSq42Nv0DEQgSbgIVgp6Sz0SKCX2CRFLFV/MAogoioyLWY3ZAi6xjcmNzlfarZ3ZnlTDF7\nmLnsvcC+Fa/3dt6rRc35a84O9efvqv08as5fc3boU/5ayuIpcDDJgSTbaAav721wprWoOX/N2aH+\n/F21n0fN+WvODv3Kv9Gj96uM5t8G3gI/aZ6tne28fwJ4STOqf2mjcw5i/pqzD0L+QTmPmvPXnH29\n8zuRoCSpVS2PoSRJG8iykCS1siwkSa0sC0lSK8tCktTKspAktbIspB6SjCSZ7mwfTTK7xuOnkuxZ\nn3TS/2VZSL2NANP/cPwUYFloIPinPKmHFVM5L9L8I/YrsASMA8+B06WUkuQQcAUY7nw+BRwBZmjm\n4PkGTAAXgJPAduARcK54AaoSloXUQ2ddgNlSynhnjZS7wBjwBpij+fF/DDykWSPjQ5JTwPFSypkk\nD4DzpZRnne8bLaV86mzfAu6UUjbztOLSsoGeolzqsyellNcASV4A+4HPNHca95t1n9hCMzfPao4l\nuQjsAEaBBTb3GhTSMstC+nvfV2z/orl+AiyUUib+dGCSIeA6cLiU8irJZWBovYJK/eYAt9TbF2Bn\nyz6LwO4kEwBJtiYZW+X4bjEsJRkGJvsdVlpP3llIPZRSPiaZSzJPM0j9fpV9fiSZBK4l2UVzTV2l\necQ0A9xI0h3gvgnMA+9o1hiQquEAtySplY+hJEmtLAtJUivLQpLUyrKQJLWyLCRJrSwLSVIry0KS\n1MqykCS1+g0Myzaq9ZssqAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x115b55198>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "simcoal_out = '/Users/tobias/GitHub/abc_modeling_course_tjarno_2017/data/constsize_obs_1_1.arp'\n",
    "alignment, num_seqs , input_dir, snp_length, alignment_length, mutation_rate, n_alleles = fscoal.convert_simcoal_output_to_fasta(simcoal_out)\n",
    "possible_combos = get_possible_combinations_for_n_sequences(num_seqs)\n",
    "aln_format = 'fasta'\n",
    "min_num_seqs = num_seqs\n",
    "outgroup = ''\n",
    "name = alignment.split('/')[-1]\n",
    "\n",
    "snp_columns = {}\n",
    "snp_positions = list(range(0,snp_length))\n",
    "snp_columns.setdefault(name,snp_positions)\n",
    "\n",
    "valid_columns = {}\n",
    "all_positions = list(range(0,alignment_length))\n",
    "valid_columns.setdefault(name,all_positions)\n",
    "\n",
    "locus_tajima_dict,locus_snp_count_dict = tajimas_estimator_per_locus(input_dir,aln_format,snp_columns,valid_columns,min_num_seqs,outgroup,possible_combos)\n",
    "tajima = locus_tajima_dict[name]\n",
    "\n",
    "print('tajima estimated log(theta) = %f'%math.log10(tajima))\n",
    "print('real log(theta) = %f'%math.log10(2*n_alleles*mutation_rate))\n",
    "\n",
    "plot_expected_heterozyosity(locus_tajima_dict,input_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "//Parameters for the coalescence simulation program : simcoal.exe\n",
      "1 samples to simulate\n",
      "//Deme sizes (haploid number of genes)\n",
      "10000\n",
      "//Sample sizes\n",
      "20\n",
      "//Growth rates\n",
      "0\n",
      "//Number of migration matrices : If 0 : No migration between demes\n",
      "0\n",
      "//Historical event: time, source, sink, migrants, new deme size, new growth rate, new migration matrix\n",
      "0 historical events\n",
      "//Number of independent chromosome\n",
      "1 0\n",
      "//Number of contiguous linkage blocks\n",
      "1\n",
      "//Per Block: Data type, No. of loci, Recombination rate to the right-side locus, plus optional parameters\n",
      "DNA 10000 0 1.0e-07 0.5\n",
      "fastsimcoal was invoked with the following command line arguments:\n",
      "bin/fsc25221 -i constsize_obs.par -n 1 \n",
      "\n",
      "Random generator seed : 201352\n",
      "\n",
      "Deme sizes\n",
      "Deme 0\t10000\n",
      "\n",
      "\n",
      "Sample sizes\n",
      "Deme 0\t20\n",
      "\n",
      "Sample ages\n",
      "Deme 0\t0\n",
      "\n",
      "Growth rates\n",
      "Deme 0\t0\n",
      "\n",
      "\n",
      "Historical events\n",
      "No historical events defined in input file ...\n",
      "\n",
      "\n",
      "Number of independent loci to simulate : 1\n",
      "with the same chromosomal structure\n",
      "\n",
      "Number of linkage blocks to simulate in structure 1: 1\n",
      "\n",
      "   10000 partially linked DNA     :\n",
      "      recombination, mutation rates and transition rates :0.0000000000   0.0000001000   0.50000\n",
      "\n",
      "fastsimcoal2 is building 1 genealogies ...\n",
      "\n",
      "Simulating 1 independent chromosomes using 12 batches and 1 threads.\n",
      "\n",
      "MRCA time for non-recombining chromosome structure 0: 7029\n",
      "\tGenealogy # 1/1\n",
      "\n",
      "Iteration 1/1 done in 0.149 secs\n",
      "\n",
      "\n",
      "\n",
      "Program execution time: 0.151 seconds\n",
      "\n",
      "\r\n",
      "\r\n",
      "             DO NOT EDIT THIS FILE BY HAND UNLESS YOU KNOW EXACTLY WHAT YOU\r\n",
      "               ARE DOING. ERRONEOUS COMPUTATIONS MAY RESULT AS A CONSEQUENCE !!!\r\n",
      "\r\n",
      "#Below are listed all parameter settings used for the choice of \r\n",
      "#ARLEQUIN computations\r\n",
      "\r\n",
      "[Setting for Calculations]\r\n",
      "\r\n",
      "#----------------------------------------------------------------------------\r\n",
      "#KEY \"TaskNumber\" sets which computations need to be performed by Arlequin\r\n",
      "#STANDARD INDICES: +1\r\n",
      "#MOLECULAR DIVERSITY: +2\r\n",
      "#MISMATCHDISTRIBution: +4\r\n",
      "#HAPLOTYPE FREQUENCIES: +8\r\n",
      "#LINKAGE DISEQUILIBRIUM: +16\r\n",
      "#HARDY-WEINBERG EQUILIBIRUM: +32\r\n",
      "#TAJIMA'S TEST: +64\r\n",
      "#EWENS-WATTERSON'S TEST: +128\r\n",
      "#CHAKRABORTY'S TEST: +256\r\n",
      "#AMOVA: +512\r\n",
      "#PAIRWISE FST: +1024\r\n",
      "#PAIRWISE GENETIC DISTANCES : +2048\r\n",
      "#PREPARE FOR AMOVA PERMUTATIONS: +4096\r\n",
      "#EXACT TEST FOR POPULATION DIFERENCES: +8192\r\n",
      "#FU FS TEST: +16384\r\n",
      "#GENOTYPE ASSIGNMENT TEST: +32768\r\n",
      "#MANTEL TEST: +65536\r\n",
      "\r\n",
      "TaskNumber=20035\r\n",
      "#----------------------------------------------------------------------------\r\n",
      "\r\n",
      "#----------------\r\n",
      "#GENERAL SETTINGS\r\n",
      "#----------------\r\n",
      "\r\n",
      "#The following 3 keys are for DNA ONLY\r\n",
      "#DeletionWeight controls the relative weight of deletions\r\n",
      "#TransitionWeight controls the relative weight of transition\r\n",
      "#TranversionWeight controls the relative weight of transversion\r\n",
      "\r\n",
      "DeletionWeight=1.000000\r\n",
      "TransitionWeight=1.000000\r\n",
      "TranversionWeight=1.000000\r\n",
      "\r\n",
      "#ComputeSumStatsWithinGroups controls if summary statistics and tests will be\r\n",
      "#computed for each group defined in the [Structure] section. If 1, then population\r\n",
      "#bsamples withing each group will be pooled into a single sample\r\n",
      "\r\n",
      "ComputeSumStatsWithinGroups=0\r\n",
      "\r\n",
      "#InferHaplotypeDefinitionFromDistanceMatrix controls if haplotype defintions are to be trusted\r\n",
      "#Put 1 if you want that the definition of the haplotyped be checked by computing distances\r\n",
      "#between all haplotypes\r\n",
      "\r\n",
      "InferHaplotypeDefinitionFromDistanceMatrix=1\r\n",
      "\r\n",
      "#EliminateRedondHaplodefs controls for the search of identical haplotypes\r\n",
      "#both within and between samples, based on their molecular diversity at the selected loci\r\n",
      "\r\n",
      "EliminateRedondHaplodefs=0\r\n",
      "\r\n",
      "#AllowedLevelOfMissingData controls the maximum level of missing data at any locus,\r\n",
      "#for the locus to be used in subsequent analyses\r\n",
      "#Note that for AMOVA, the frequency of missing data is computed over all populations\r\n",
      "#used in the genetic structure, while for intra-population computations loci are examined\r\n",
      "#separately in each population sample\r\n",
      "\r\n",
      "AllowedLevelOfMissingData=1.000000\r\n",
      "\r\n",
      "#GameticPhaseIsKnown controls if the gametic phase is assumed to be known,\r\n",
      "#or if we deal with unphased genotypic data\r\n",
      "#This is an obsolete setting, since it is the value present in the project file that prevails\r\n",
      "\r\n",
      "GameticPhaseIsKnown=0\r\n",
      "\r\n",
      "#KeepNullDistrib controls if the null distributions of AMOVA variance components and F-statistics\r\n",
      "#should be ouptut in dedicated result files\r\n",
      "\r\n",
      "KeepNullDistrib=0\r\n",
      "\r\n",
      "#-----------------------------------------------\r\n",
      "#SETTINGS FOR TEST OF HARDY-WEINBERG EQUILIBRIUM\r\n",
      "#-----------------------------------------------\r\n",
      "\r\n",
      "#MakeHWExactTest controls if exact test of HWE neEds to be performed (1: yes, 0: no)\r\n",
      "\r\n",
      "MakeHWExactTest=0\r\n",
      "\r\n",
      "#HardyWeinbergTestType: DEPRECATED, replaced by key TypeOfTestHW\r\n",
      "#                       0 for locus by locus test, \r\n",
      "#                       1 when phase is known considers haplotype as a locus\r\n",
      "\r\n",
      "HardyWeinbergTestType=0\r\n",
      "\r\n",
      "#TypeOfTestHW: 0 - Test association at the Locus level, 1 - Test association at the Haplotype level,\r\n",
      "#              2 - Test association at the Locus and Haplotype levels\r\n",
      "\r\n",
      "TypeOfTestHW=0\r\n",
      "\r\n",
      "#MarkovChainStepsHW controls the number of steps in Markov chain for HWE exact test\r\n",
      "\r\n",
      "MarkovChainStepsHW=1000000\r\n",
      "\r\n",
      "#MarkovChainDememorisationStepsHW controls the number of burnin steps in Markov <n#chain for HWE exact test\r\n",
      "\r\n",
      "MarkovChainDememorisationStepsHW=100000\r\n",
      "\r\n",
      "#PrecisionOnPValueHW controls the required precision on the estimated p-value of the HWE test\r\n",
      "#OBSOLETE FEATURE LET THE KEY VALUE TO ZERO\r\n",
      "\r\n",
      "PrecisionOnPValueHW=0.000000\r\n",
      "\r\n",
      "#SignificanceLevelHW controls the minimum p-value of a HWE test to be flagged\r\n",
      "#as significant in a summary table to be output in the result file\r\n",
      "\r\n",
      "SignificanceLevelHW=2\r\n",
      "\r\n",
      "#-------------------------------------------\r\n",
      "#SETTINGS FOR TEST OF LINKAGE DISEQUILIBRIUM\r\n",
      "#-------------------------------------------\r\n",
      "\r\n",
      "#LinkageDisequilibriumTestType: (OBSOLETE SETTING)\r\n",
      "LinkageDisequilibriumTestType=0\r\n",
      "\r\n",
      "#MakeExactTestLD controls if test of LD needs to be performed\r\n",
      "#irrespective of the fact that gamwtic phase is knwn or not\r\n",
      "#If the gametic phase is known, an exact test will be performed, and\r\n",
      "#if it is not, a likelihood-ratio will be done\r\n",
      "\r\n",
      "MakeExactTestLD=0\r\n",
      "\r\n",
      "#MarkovChainDememorisationStepsLD controls the number of burnin steps at the\r\n",
      "#beginning of the Markov chain (when gametic phae is known)\r\n",
      "\r\n",
      "MarkovChainDememorisationStepsLD=1000\r\n",
      "\r\n",
      "#MarkovChainStepsLD controls how many steps are to be performed in the Markov chain\r\n",
      "#after the burnin period, when the gametic phase is known, or it is the number of permutations\r\n",
      "#to be performed to test the p-value of the likelihood-ratio test when the gametic phase\r\n",
      "#is unknown\r\n",
      "\r\n",
      "MarkovChainStepsLD=10000\r\n",
      "\r\n",
      "#PrecisionOnPValueLD controls the precision required for the \r\n",
      "#p-value of the LD exact test (OBSOLETE SETTING)\r\n",
      "\r\n",
      "PrecisionOnPValueLD=0.000000\r\n",
      "\r\n",
      "#SignificanceLevelLD controls the p-values under which the LD test is assumed \r\n",
      "#to be significant. This is used to produce a summary table of the results in theresult file\r\n",
      "\r\n",
      "SignificanceLevelLD=0.050000\r\n",
      "\r\n",
      "#PrintFlagHistogramLD controls if a file with a summary of the LD test results needs ot be produced\r\n",
      "\r\n",
      "PrintFlagHistogramLD=0\r\n",
      "\r\n",
      "#InitialCondEMLD controls the number of different EM algorithm runs to\r\n",
      "#perform for the LD likelihood ratio test when the gametic phase is not known\r\n",
      "\r\n",
      "InitialCondEMLD=2\r\n",
      "\r\n",
      "#ComputeDvalues controls if coefficients D and D' needs to be computed, \r\n",
      "#Note that this option is only valid if gametic phase is known\r\n",
      "\r\n",
      "ComputeDvalues=0\r\n",
      "\r\n",
      "#------------------------------\r\n",
      "#SETTINGS FOR DIVERSITY INDICES\r\n",
      "#------------------------------\r\n",
      "\r\n",
      "#ComputeStandardDiversityIndices controls if standard diversity indices are to be computed\r\n",
      "\r\n",
      "ComputeStandardDiversityIndices=1\r\n",
      "\r\n",
      "#ComputeAllAlleleFreqs controls if allele frequencies are computed at all loci\r\n",
      "\r\n",
      "ComputeAllAlleleFreqs=0\r\n",
      "\r\n",
      "#DistanceMethod controls the choice of the distance to be computed between haplotypes\r\n",
      "#DNA:      DIFF=0, JUKES_CANTOR=1, KIMURA_2P=2, P_DIST=3, TAJIMA_NEI=4, TAMURA=5, TAMURA_NEI=6\r\n",
      "#MICROSAT: COUNT_DIFF=0 (FST), MICRO_SQR_DIFF=1 (RST)\r\n",
      "#STANDARD: 0 (default)\r\n",
      "#RFLP:     RFLP_DIFF=0, RFLP_P_DIST=1\r\n",
      "\r\n",
      "DistanceMethod=0\r\n",
      "\r\n",
      "#GammaAValue controls the value of the shape parameter of gamma distributed mutaiton rates (for DNA only)\r\n",
      "\r\n",
      "GammaAValue=0.000000\r\n",
      "\r\n",
      "#ComputeTheta controls if theta->4Nu parameters are to be computed\r\n",
      "#Bit values : +1: Theta(Hom), +2: Theta(S), +4: Theta(k), +8: Theta(pi)\r\n",
      "\r\n",
      "ComputeTheta=15\r\n",
      "\r\n",
      "#PrintHaplDistMat controls if a distance matrix between haplotypes needs to be output in result file\r\n",
      "\r\n",
      "PrintHaplDistMat=0\r\n",
      "\r\n",
      "#PrintPopDistMat has the same effect as \"PrintHaplDistMat\" above and should have the same value\r\n",
      "\r\n",
      "PrintPopDistMat=0\r\n",
      "\r\n",
      "#PrintMinSpannNetworkPop controls if a Minimum Spaning Tree computed from the\r\n",
      "#matrix of pairwise distances between haplotypes needs to be output in the result file\r\n",
      "\r\n",
      "PrintMinSpannNetworkPop=0\r\n",
      "\r\n",
      "#----------------------------------\r\n",
      "#SETTINGS FOR MISMATCH DISTRIBUTION\r\n",
      "#----------------------------------\r\n",
      "\r\n",
      "#MismatchDemogExp specifies if the parameters of a pure demographic expansions\r\n",
      "#will be computed from the mismatch distribution\r\n",
      "\r\n",
      "MismatchDemogExp=0\r\n",
      "\r\n",
      "#MismatchSpatialExp controls if one attempts to estimate parameters of a spatial range expansion\r\n",
      "\r\n",
      "MismatchSpatialExp=0\r\n",
      "\r\n",
      "#MismatchDistanceMethod controls the distance to be computed between haplotypes\r\n",
      "#OBSOLETE: It is now always set to zero\r\n",
      "\r\n",
      "MismatchDistanceMethod=0\r\n",
      "\r\n",
      "#MismatchGammaAValue controls the value of the shape parameter of gamma distributed \r\n",
      "#mutation rates (for DNA only)\r\n",
      "#OBSOLETE: Always set to zero\r\n",
      "\r\n",
      "MismatchGammaAValue=0.000000\r\n",
      "\r\n",
      "#NumBootExpDem controls the number of bootstrap replicates used to compute the confidence\r\n",
      "#intervals for parameters and p-value of the demographic and/or spatial expansion hypotheses\r\n",
      "\r\n",
      "NumBootExpDem=100\r\n",
      "\r\n",
      "#MismatchSpatialExpNumParams controls the type of range expansion\r\n",
      "# 3: Three parameters are used: Theta=2Nu, M=2Nm, tau=2Tu\r\n",
      "\r\n",
      "MismatchSpatialExpNumParams=3\r\n",
      "\r\n",
      "#-------------------------------------------\r\n",
      "#SETTINGS FOR HAPLOTYPE FREQUENCY ESTIMATION\r\n",
      "#-------------------------------------------\r\n",
      "\r\n",
      "#EM algorithm specific settings\r\n",
      "#------------------------------\r\n",
      "\r\n",
      "\r\n",
      "#ComputeAllHaplotypesEM controls if we need to estimate haplotype frequencies through the EM algorithm\r\n",
      "\r\n",
      "ComputeAllHaplotypesEM=0\r\n",
      "\r\n",
      "#ComputeAllAllelesEM controls if we need to estimate all allele frequencies with the EM algorithm\r\n",
      "#The EM algorithm is here mostly useful if there are recessive alleles\r\n",
      "\r\n",
      "ComputeAllAllelesEM=0\r\n",
      "\r\n",
      "#InitialConditionsEM controls the number of times the EM algorithm has to be performed\r\n",
      "#from different initial conditions in attempting to find the global maximu-likelihood\r\n",
      "\r\n",
      "InitialConditionsEM=50\r\n",
      "\r\n",
      "#MaximumNumOfIterationsEM controls the maximum number of iterations of the EM algorithm per run\r\n",
      "\r\n",
      "MaximumNumOfIterationsEM=1000\r\n",
      "\r\n",
      "#RecessiveAllelesEM controls if recessive alleles are to be considered in homozygotes\r\n",
      " \r\n",
      "RecessiveAllelesEM=0\r\n",
      "\r\n",
      "#CompactHaplotypeDataBaseEM controls if loci that are monomorphic are to be removed from the haplotypes\r\n",
      "#which marginally speeds up the computation\r\n",
      "#Note that this parameter is now OBSOLETE and should always be set to zero\r\n",
      "\r\n",
      "CompactHaplotypeDataBaseEM=0\r\n",
      "\r\n",
      "#NumBootstrapReplicatesEM controls the number of bootstraps to perform in order to get estimates\r\n",
      "#of standard deviations of the haplotype frequencies. Note thats.d. estimation takes a LOT of time\r\n",
      "\r\n",
      "NumBootstrapReplicatesEM=0\r\n",
      "\r\n",
      "#NumInitCondBootstrapEM controls the number of random initial consitions for the EM,\r\n",
      "#when performing the bootstrap s.d. estimation, much like InitialConditionsEM, but it should have\r\n",
      "#typically alower value than that of InitialConditionsEM to save up time\r\n",
      "\r\n",
      "NumInitCondBootstrapEM=10\r\n",
      "\r\n",
      "#ComputeAllSubHaplotypesEM controls if we estimate all possible two-locus haplotype frequencies\r\n",
      "\r\n",
      "ComputeAllSubHaplotypesEM=0\r\n",
      "\r\n",
      "#EMCheckPhenRedundancy controls if phenotypes need to be checked to see if they are all different.\r\n",
      "#If, yes, it removes some phenotypes and update their frequency, if needed\r\n",
      "\r\n",
      "EMCheckPhenRedundancy=0\r\n",
      "\r\n",
      "#EpsilonValue is a threshold value: when the sum of haplotype frequency differences \r\n",
      "#between consecutive iterations is larger than epsilon, the EM algorithm is believed to have converged,\r\n",
      "#and stops. Values of at least 1E-7 are in order\r\n",
      "\r\n",
      "EpsilonValue=1.000000e-07\r\n",
      "\r\n",
      "#FrequencyThreshold is the minimum frequency need to be reached by a haplotype to be listed in the result file\r\n",
      "\r\n",
      "FrequencyThreshold=1.000000e-05\r\n",
      "\r\n",
      "#EMEstimateInbreedingCoeff controls if we need to estimate the inbreeding coefficient together with the\r\n",
      "#the haplotype frequencies using the EM algorithm\r\n",
      "\r\n",
      "EMEstimateInbreedingCoeff=0\r\n",
      "\r\n",
      "#EMInbreedingFixedVal sets a predefined arbitrary value of inbreeding, which is taken into\r\n",
      "#account when estimating haplotype frequencies through the EM algorithm\r\n",
      "\r\n",
      "EMInbreedingFixedVal=0.000000\r\n",
      "\r\n",
      "#EMUseZipper controls if we use the zipper version of the EM algorithm (unpublished)\r\n",
      "#The zipper version works by computing first 2-locus haplotype frequencies, and then progressively\r\n",
      "#adding other loci in a random order and estimating haplotype frequencies on a larger number of loci\r\n",
      "#until all loci have been included. This procedure is much quicker than the conventional EM algorithm\r\n",
      "#as a much more restricted set of haplotypes have to be considered.\r\n",
      "#It can thus be used on a much largern number of loci than the conventinal EM algorithm\r\n",
      "\r\n",
      "EMUseZipper=0\r\n",
      "\r\n",
      "#EMLociRandOrder specifies if loci are to be added in a random order for the zipper version of the EM\r\n",
      "\r\n",
      "EMLociRandOrder=0\r\n",
      "\r\n",
      "#EMNumZipRandOrders specifies how many random orders have to be considered assuming EMLociRandOrder=1\r\n",
      "\r\n",
      "EMNumZipRandOrders=0\r\n",
      "\r\n",
      "#ELB algorithm specific settings\r\n",
      "#-------------------------------\r\n",
      "\r\n",
      "\r\n",
      "#GibbsInferGamPhase controls if we estimate individual gametic phase and haplotype frequencies\r\n",
      "#via the pseudo-Gibbs sampling implemented in the ELB algorithm\r\n",
      "\r\n",
      "GibbsInferGamPhase=0\r\n",
      "\r\n",
      "#GibbsBurnIn defines the number of the Gibbs steps in the burnin\r\n",
      "\r\n",
      "GibbsBurnIn=100000\r\n",
      "\r\n",
      "#GibbsSamplingInterval defines the number of Gibbs steps between each consecutive samples\r\n",
      "#in the Markov chain\r\n",
      "\r\n",
      "GibbsSamplingInterval=500\r\n",
      "\r\n",
      "#GibbsNumSamples defines how many samples we need to get in order to build the posterior distribution\r\n",
      "#of the individual gametic phases\r\n",
      "\r\n",
      "GibbsNumSamples=2000\r\n",
      "\r\n",
      "#GibbsAlphaInitValue is the value of the alpha parameter of the Dirichlet prior distribution\r\n",
      "#for haplotype frequencies. A value of 0.01 has been shown to work best in all cases.\r\n",
      "\r\n",
      "GibbsAlphaInitValue=0.010000\r\n",
      "\r\n",
      "#GibbsEpsilonValue is the value of the epsilon parameter, used to take into account the frequency\r\n",
      "#of haplotypes similar to those possible for a given unphase individual. A value of 0.01 works well for\r\n",
      "#SNPS, and a larger value of 0.1 works well for microsatellite data\r\n",
      "\r\n",
      "GibbsEpsilonValue=0.010000\r\n",
      "\r\n",
      "#GibbsGamma is the gamma parameter of the ELB algorithm, used to avoid dynamic windows to grow too large,\r\n",
      "#when the epsilon is small and if there is a strong linkage disequilibrium.\r\n",
      "\r\n",
      "GibbsGamma=0.000000\r\n",
      "\r\n",
      "#GibbsHetSiteInfluenceZone controls the number of loci to be taken into account \r\n",
      "#around each heterozygous site, to dress a list, for each individual, of the loci\r\n",
      "#that must be take into account to build haplotypes.\r\n",
      "#Note that with a value of zero, gametic phase is computed only on heterozygotes sites, while\r\n",
      "#a negative value(e.g. -1) implies that haplotypes will be defined on all loci.\r\n",
      "\r\n",
      "GibbsHetSiteInfluenceZone=5\r\n",
      "\r\n",
      "#GibbsPercentageOfRecSteps controls the fraction of steps in Gibbs sampling that \r\n",
      "#correspond to recombination steps around the focal heterozygous locus\r\n",
      "\r\n",
      "GibbsPercentageOfRecSteps=0.000000\r\n",
      "\r\n",
      "#GibbsOutputGamPhase controls if we output individual gametic phases in *.arp files \r\n",
      "#for every sample of the ELB algorithm\r\n",
      "\r\n",
      "GibbsOutputGamPhase=0\r\n",
      "\r\n",
      "#------------------\r\n",
      "#SETTINGS FOR AMOVA\r\n",
      "#------------------\r\n",
      "\r\n",
      "#HaplAMOVA controls if the AMOVA computations need to be performed globally,\r\n",
      " for all loci at the same time, as for haplotypes\r\n",
      "\r\n",
      "HaplAMOVA=0\r\n",
      "\r\n",
      "#LocByLocAMOVA controls if the AMOVA computations need to be performed separately for each locus.\r\n",
      "\r\n",
      "LocByLocAMOVA=1\r\n",
      "\r\n",
      "#PopSpecificFST controls if population specific FST indices need to be computed.\r\n",
      "\r\n",
      "PopSpecificFST=0\r\n",
      "PopSpecificFIS=1\r\n",
      "\r\n",
      "#TestSignificanceAMOVA should be set to 1 to perfrom significance testing of the variance components\r\n",
      "#OBSOLETE settings (could probably be removed), because significance testing is \r\n",
      "#controlled by NumPermutationsAMOVA below\r\n",
      "\r\n",
      "TestSignificanceAMOVA=1\r\n",
      "\r\n",
      "#NumPermutationsAMOVA controls if variance components are tested by permutations, \r\n",
      "#and sets the number of permutations to be performed\r\n",
      "NumPermutationsAMOVA=0\r\n",
      "\r\n",
      "#ComputeConventionalFST controls if conventional FSt are to be computed as opposed to\r\n",
      "#F-statistics taking into account molecular diversity. Conventional FST do only depend on\r\n",
      "#haplotype frequencies, and do not take into account the allelci composition of the haplotypes\r\n",
      "\r\n",
      "ComputeConventionalFST=0\r\n",
      "\r\n",
      "#IncludeIndividualLevel controls if FIS statistics and associated variance components\r\n",
      "#should be computed\r\n",
      "\r\n",
      "IncludeIndividualLevel=1\r\n",
      "\r\n",
      "#ComputeDistanceMatrixAMOVA controls if a matrix of pairwise Euclidian distances needs to be computed\r\n",
      "#and used for AMOVA computaitons\r\n",
      "\r\n",
      "ComputeDistanceMatrixAMOVA=1\r\n",
      "\r\n",
      "#DistanceMethodAMOVA controls if a matrix of euclidian distances should be computed\r\n",
      "# between all pairs of haplotypes prior to AMOVA computations\r\n",
      "\r\n",
      "DistanceMethodAMOVA=0\r\n",
      "\r\n",
      "#GammaAValueAMOVA is the shape parameter of gamma distributed mutation rates.\r\n",
      "#A value of zero means no heterogeneity of mutation rates. Values between zero and one\r\n",
      "#and increasingly smaller imply increasing departure from mutation rate homogeneity\r\n",
      "#For DNA distances only\r\n",
      "\r\n",
      "GammaAValueAMOVA=0.000000\r\n",
      "\r\n",
      "#PrintDistanceMatrix controls if the matrix of pairwise distance between haplotypes should\r\n",
      "#be output in the result file\r\n",
      "\r\n",
      "PrintDistanceMatrix=0\r\n",
      "\r\n",
      "#PrintMinSpannNetworkGlob controls if a Minimum Spanning Network need sto be computed from a matrix\r\n",
      "#or pairwise distances between all pairs of haplotypes found in the population samples included in the\r\n",
      "#current genetic structure\r\n",
      "\r\n",
      "PrintMinSpannNetworkGlob=0\r\n",
      "\r\n",
      "#-------------------------------------------\r\n",
      "#SETTINGS FOR PAIRWISE DISTANCE COMPUTATIONS\r\n",
      "#-------------------------------------------\r\n",
      "\r\n",
      "#TestSignificancePairewiseFST controls if pairwise distances should be tested by permutations\r\n",
      "\r\n",
      "TestSignificancePairewiseFST=0\r\n",
      "\r\n",
      "#NumPermutationsFST is the number of permutations to be perfromed to test the significance\r\n",
      "#of pairwise distances between populations\r\n",
      "\r\n",
      "NumPermutationsFST=0\r\n",
      "\r\n",
      "#ComputePairwiseFST controls if FST's should be computed between all pairs of populations\r\n",
      "\r\n",
      "ComputePairwiseFST=1\r\n",
      "\r\n",
      "#ComputeDeltaMuSqr controls if delta-mu square statistics is computed between all pairs of populations\r\n",
      "\r\n",
      "ComputeDeltaMuSqr=1\r\n",
      "\r\n",
      "#FSTSignifLevel is the p-value under which pairwise distances are assumed to be significant\r\n",
      "#and reported as such in the result file\r\n",
      "\r\n",
      "FSTSignifLevel=0.050000\r\n",
      "\r\n",
      "#PrintFstVals controls if pairwise FST values need to be printed in the result file\r\n",
      "\r\n",
      "PrintFstVals=1\r\n",
      "\r\n",
      "#PrintConcestryCoeff controls if Reynold's distances need to be printed in the result file\r\n",
      "\r\n",
      "PrintConcestryCoeff=0\r\n",
      "\r\n",
      "#PrintSlatkinsDist controls if Slatkin's distances need to be printed in the result file\r\n",
      "\r\n",
      "PrintSlatkinsDist=0\r\n",
      "\r\n",
      "#PrintMissIntermatchs controls average number of pairwise differences between populations\r\n",
      "#need to be computed and reported in the result file\r\n",
      "\r\n",
      "PrintMissIntermatchs=1\r\n",
      "\r\n",
      "#UnequalPopSizeDiv controls if the distances taking into account diferences between population sizes\r\n",
      "#need to be computed and reported in the result file\r\n",
      "\r\n",
      "UnequalPopSizeDiv=0\r\n",
      "\r\n",
      "#---------------------------------------\r\n",
      "#SETTINGS FOR POPULATION DIFFERENTIATION\r\n",
      "#---------------------------------------\r\n",
      "\r\n",
      "#NumPermutPopDiff is the number of steps in the Markov chain for the exact test\r\n",
      "#of population differention based on haplotype frequency differences between populations\r\n",
      "\r\n",
      "NumPermutPopDiff=0\r\n",
      "\r\n",
      "#NumDememoPopDiff is the number of steps in the burnin of the markov chain used for the\r\n",
      "\r\n",
      "exact test of popualtion differentiation\r\n",
      "\r\n",
      "NumDememoPopDiff=0\r\n",
      "\r\n",
      "#PrecProbPopDiffcontrols the required precision on the estimated p-value of the exact test\r\n",
      "#of population differentiation. OBSOLETE FEATURE LET THE KEY VALUE TO ZERO\r\n",
      "\r\n",
      "PrecProbPopDiff=0.000000\r\n",
      "\r\n",
      "#PrintHistoPopDiff controls if a summary table of siginificant differences beteen populations\r\n",
      "#nshould be printed in the result file\r\n",
      "\r\n",
      "PrintHistoPopDiff=1\r\n",
      "\r\n",
      "#SignLevelPopDiffis the p-value under which population differences are assumed to be significant\r\n",
      "#and reported as such in the result file\r\n",
      "\r\n",
      "SignLevelPopDiff=0.050000\r\n",
      "\r\n",
      "#--------------------------------------------------------------------------------------\r\n",
      "#SETTINGS FOR DETECTION OF SELECTED LOCI FROM COMPARISON BETWEEN FST AND HETEROZYGOSITY\r\n",
      "#--------------------------------------------------------------------------------------\r\n",
      "\r\n",
      "#DetectSelectedLoci controls if the Beaumont and Nichols (1996) test of selection (FDIST2) will be performed\r\n",
      "\r\n",
      "DetectSelectedLoci=0\r\n",
      "\r\n",
      "#UseHierarchicalIslandModel controls if a hierarchical island model should be used to generate the null distribution of FST\r\n",
      "\r\n",
      "UseHierarchicalIslandModel=0\r\n",
      "\r\n",
      "#NumberOfGroupsForFDist2 is the number of groups in the simulated hierarchical island model\r\n",
      "\r\n",
      "NumberOfGroupsForFDist2=0\r\n",
      "\r\n",
      "#NumberOfGroupsForFDist2 is the number of demes (per group) in the simulated island model\r\n",
      "\r\n",
      "NumberOfDemesForFDist2=0\r\n",
      "\r\n",
      "#NumberOfSimsForFDist2 is the number of coalescent simulations to be performed\r\n",
      "\r\n",
      "NumberOfSimsForFDist2=0\r\n",
      "\r\n",
      "#UseSpecificAveFSTForFDist2 controls if a specific value of FST has to be used for the coalescent simulations\r\n",
      "#If set to zero, the FST value will be computed from the current genetic structure\r\n",
      "\r\n",
      "UseSpecificAveFSTForFDist2=0\r\n",
      "\r\n",
      "#SpecificFSTForFDist2 is the average target FST for the coalescent simulations under a finite island model\r\n",
      "\r\n",
      "SpecificFSTForFDist2=0.000000\r\n",
      "\r\n",
      "#TargetHetLowBound is the target LOW bound for simulated heterozagosity in the coalescent simulations under a finite island model\r\n",
      "#Target simulated heterozygostiy is dawn as uniform between low_bound and high_bound and theta value is taken from the relation Het=theta/(1+theta) \r\n",
      "\r\n",
      "TargetHetLowBound=0.000000\r\n",
      "\r\n",
      "#TargetHetHighBound is the target HIGH bound for simulated heterozagosity in the coalescent simulations under a finite island model\r\n",
      "#Target simulated heterozygostiy is dawn as uniform between low_bound and high_bound and theta value is taken from the relation Het=theta/(1+theta) \r\n",
      "\r\n",
      "TargetHetHighBound=0.000000\r\n",
      "\r\n",
      "#IterativeProcForFDist2 controls if we use an iterative procedure to detect loci under selection\r\n",
      "\r\n",
      "IterativeProcForFDist2=0\r\n",
      "\r\n",
      "#-----------------------------\r\n",
      "#SETTINGS FOR NEUTRALITY TESTS\r\n",
      "#-----------------------------\r\n",
      "\r\n",
      "#EwensWattersonHomozygosityTest controls if the Ewens-watterson neutraltiy test should be performed\r\n",
      "\r\n",
      "EwensWattersonHomozygosityTest=0\r\n",
      "\r\n",
      "#NumIterationsNeutralityTests is the number of simulations to be performed for the\r\n",
      "#neutrality tests based on the infinite-allele model\r\n",
      "\r\n",
      "NumIterationsNeutralityTests=0\r\n",
      "\r\n",
      "#NumSimulFuTestis the number of coalescent simulations to be performed for the\r\n",
      "#neutrality tests based on the infinite-site model\r\n",
      "\r\n",
      "NumSimulFuTest=0\r\n",
      "\r\n",
      "#------------------------\r\n",
      "#SETTINGS FOR MANTEL TEST\r\n",
      "#------------------------\r\n",
      "\r\n",
      "#NumPermMantel is the number of permutations to be perform to get the significance of\r\n",
      "#the correlation between different matrices\r\n",
      "\r\n",
      "NumPermMantel=0\r\n",
      "List of summary statistics to compute (# sign means that these summary statisticss will not be computed)\r\n",
      "-------------------------------------\r\n",
      "POP_LEVEL\r\n",
      "#GROUP_LEVEL\r\n",
      "#ALL_K\r\n",
      "#ALLSD_K\r\n",
      "#MEAN_K\r\n",
      "#SD_K\r\n",
      "#TOT_K\r\n",
      "ALL_H\r\n",
      "#ALLSD_H\r\n",
      "#MEAN_H\r\n",
      "#SD_H\r\n",
      "#TOT_H\r\n",
      "ALL_S\r\n",
      "PRIVATE_S\r\n",
      "#MEAN_S\r\n",
      "#SD_S\r\n",
      "#TOT_S\r\n",
      "ALL_TAJIMAD\r\n",
      "#MEAN_TAJIMAD\r\n",
      "#SD_TAJIMAD\r\n",
      "ALL_FUFS\r\n",
      "#MEAN_FUFS\r\n",
      "#SD_FUFS\r\n",
      "ALL_PI\r\n",
      "#MEAN_PI\r\n",
      "#SD_PI\r\n",
      "PAIRWISE_PI\r\n",
      "#PAIRWISE_DMUSQR\r\n",
      "#ALL_GW\r\n",
      "#ALLSD_GW\r\n",
      "#MEAN_GW\r\n",
      "#SD_GW\r\n",
      "#TOT_GW\r\n",
      "#ALL_NGW\r\n",
      "#ALLSD_NGW\r\n",
      "#MEAN_NGW\r\n",
      "#SD_NGW\r\n",
      "#ALL_RANGE\r\n",
      "#ALLSD_RANGE\r\n",
      "#MEAN_RANGE\r\n",
      "#SD_RANGE\r\n",
      "#TOT_RANGE\r\n",
      "#GLOBAL_FSTAT\r\n",
      "PAIRWISE_FST\r\n",
      "\r\n",
      "// Priors and rules file\n",
      "// *********************\n",
      "\n",
      "[PARAMETERS]\n",
      "//#isInt? #name\t#dist.#min\t#max\t\n",
      "//all N are in number of diploid individuals\n",
      "0\tLOG10_N_NOW\tunif\t2\t6 output\n",
      "0\tLOG10_MUTATION_RATE\tunif\t-9\t-6 output\n",
      "\n",
      "[COMPLEX PARAMETERS]\n",
      "1\tN_NOW = pow10( LOG10_N_NOW ) output\n",
      "0\tMUTATION_RATE = pow10( LOG10_MUTATION_RATE ) output\n",
      "\n",
      "//Parameters  for the coalescence simulation program : simcoal.exe\n",
      "1 samples to simulate\n",
      "//Deme sizes (haploid number of genes)\n",
      "N_NOW\n",
      "//Sample sizes\n",
      "20\n",
      "//Growth rates\n",
      "0\n",
      "//Number of migration matrices : If 0 : No migration between demes\n",
      "0\n",
      "//Historical event: time, source, sink, migrants, new deme size, new growth rate, new migration matrix\n",
      "0 historical events\n",
      "//Number of independent chromosome\n",
      "1 0\n",
      "//Number of contiguous linkage blocks\n",
      "1\n",
      "//Per Block: Data type, No. of loci, Recombination rate to the right-side locus, plus optional parameters\n",
      "DNA 10000 0 MUTATION_RATE 0.5\n",
      "\n",
      "\n",
      "\n",
      "//----------------------------------------------------------------------\n",
      "//Inputfile to simulate contant size population with N and mu\n",
      "//----------------------------------------------------------------------\n",
      "task simulate\n",
      "\n",
      "estName constsize.est\n",
      "outName sims_constsize\n",
      "\n",
      "obsName constsize.obs\n",
      "simDataName constsize-temp_1_1.arp\n",
      "numSims 10000\n",
      "simProgram bin/fsc25221\n",
      "simInputName constsize.par\n",
      "simArgs -i SIMINPUTNAME -n 1 -q\n",
      "sumStatProgram bin/arlsumstat3522_64bit\n",
      "sumStatArgs SIMDATANAME SSFILENAME 0 1\n",
      "logFile simulation.log\n",
      "\n",
      "verbose\n",
      "childOutputName childoutput.txt\n",
      "\n",
      " ABCtoolbox 2.0 \n",
      "****************\n",
      "- Reading inputfile 'constsize.input' ... done!\n",
      "- Writing log to 'simulation.log'\n",
      "- Initializing random generator ... done with seed 67158664!\n",
      "- General settings:\n",
      "   - Suffix for temporary output files: -temp\n",
      "   - Prefix for output files: sims_constsize\n",
      "   - All Statistics are written in one single Outputfile\n",
      "   - Always performing 1 run(s) per parameter vector\n",
      "   - Output of child processes is written to 'childoutput.txt'\n",
      "   - Will abort sampling when external programs return errors.\n",
      "- Initializing simulations:\n",
      "   - Reading priors and rules from 'constsize.est' ... done!\n",
      "      -> 2 parameters\n",
      "      -> 2 complex parameters\n",
      "      -> 0 rules\n",
      "   - Only executing simulation programs if required model parameters changed.\n",
      "   - Initializing simulation programs:\n",
      "      1)  Initializing 'bin/fsc25221':\n",
      "         - Input file(s) and corresponding name(s) after parsing:\n",
      "            - 'constsize.par' -> 'constsize-temp.par'\n",
      "         - Using simulation program 'bin/fsc25221'\n",
      "            - Command line to be used: bin/fsc25221 -i constsize-temp.par -n 1 -q\n",
      "               -> No parameters among arguments\n",
      "         - Launching a test simulation (with simulation number -1):\n",
      "            - Creating first input file(s)  ... done!\n",
      "            - Command lines to be used:\n",
      "               - Simulation program: bin/fsc25221 -i constsize-temp.par -n 1 -q\n",
      "            - Performing simulation ... done!\n",
      "- Initializing summary statistics from simulation output:\n",
      "   - Reading obs files:\n",
      "      - Reading observed data file 'constsize.obs' ... done!\n",
      "         -> 1 data sets with 6 statistics each.\n",
      "      -> Created a total of 1 data objects\n",
      "   - Initializing data objects:\n",
      "      1)  Data object 1:\n",
      "         - Expecting simulated summary statistics in file 'summary_stats-temp.txt'.\n",
      "         - Expecting to find the simulate data in 'constsize-temp_1_1.arp'\n",
      "            -> located in 'constsize-temp/constsize-temp_1_1.arp'\n",
      "         - Using program 'bin/arlsumstat3522_64bit' to calculate summary statistics\n",
      "            - Command line to be used: bin/arlsumstat3522_64bit constsize-temp/constsize-temp_1_1.arp summary_stats-temp.txt 0 1\n",
      "         - Execute on data from test simulation ... done!\n",
      "         - Reading simulated summary statistics from 'summary_stats-temp.txt' ... done\n",
      "            -> This file is located in the working directory.\n",
      "            -> Read 6 Data\n",
      "- Performing a standard simulation run:\n",
      "   - Performing a total of 10000 simulations\n",
      "\n",
      "- Program terminated in 0 min!\n",
      "\n",
      "\n",
      "//----------------------------------------------------------------------\n",
      "//ABCtoolbox input file to do estimations of N and mu\n",
      "//----------------------------------------------------------------------\n",
      "task estimate\n",
      "estimationType standard\n",
      "\n",
      "//observed data\n",
      "obsName constsize.obs\n",
      "\n",
      "//simulated data\n",
      "simName sims_constsize_sampling1.txt\n",
      "params 2-3 \n",
      "outputPrefix ABC_estimation_constsize_\n",
      "\n",
      "//Rejection settings\n",
      "numRetained 2000\n",
      "maxReadSims 100000\n",
      "standardizeStats 1\n",
      "writeRetained 0\n",
      "\n",
      "//Posterior estimation settings\n",
      "diracPeakWidth 0.02\n",
      "posteriorDensityPoints 200\n",
      "pruneCorrelatedStats\n",
      "\n",
      "//settings for joint posterior\n",
      "jointPosteriors LOG10_MUTATION_RATE,LOG10_N_NOW\n",
      "jointPosteriorDensityPoints 100\n",
      "\n",
      "//output settings\n",
      "logFile estimation.log\n",
      "verbose\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " ABCtoolbox 2.0 \n",
      "****************\n",
      "- Reading inputfile 'estimate.input' ... done!\n",
      "- Writing log to 'estimation.log'\n",
      "- Initializing random generator ... done with seed 67171707!\n",
      "- Reading observed data file 'constsize.obs' ... done!\n",
      "   -> 1 data sets with 6 statistics each.\n",
      "- Reading files with simulations:\n",
      "   - Reading up to 100000 simulations from file 'sims_constsize_sampling1.txt' ...\r",
      "   - Reading up to 100000 simulations from file 'sims_constsize_sampling1.txt' ... (1%)\r",
      "   - Reading up to 100000 simulations from file 'sims_constsize_sampling1.txt' ... (2%)\r",
      "   - Reading up to 100000 simulations from file 'sims_constsize_sampling1.txt' ... (3%)\r",
      "   - Reading up to 100000 simulations from file 'sims_constsize_sampling1.txt' ... (4%)\r",
      "   - Reading up to 100000 simulations from file 'sims_constsize_sampling1.txt' ... (5%)\r",
      "   - Reading up to 100000 simulations from file 'sims_constsize_sampling1.txt' ... (6%)\r",
      "   - Reading up to 100000 simulations from file 'sims_constsize_sampling1.txt' ... (7%)\r",
      "   - Reading up to 100000 simulations from file 'sims_constsize_sampling1.txt' ... (8%)\r",
      "   - Reading up to 100000 simulations from file 'sims_constsize_sampling1.txt' ... (9%)\r",
      "   - Reading up to 100000 simulations from file 'sims_constsize_sampling1.txt' ... (10%)\r",
      "   - Reading up to 100000 simulations from file 'sims_constsize_sampling1.txt' ... done!   \n",
      "      -> 10000 simulations with 2 parameters and 6 statistics each.\n",
      "   - Pruning highly correlated statsitics:\n",
      "      - Calculate pairwise correlations ...done!\n",
      "      - Found correlations above chosen threshold (0.95): \n",
      "         - Removing statistics 'S_1' (correlation with 'prS_1' is 1)\n",
      "         - Removing statistics 'prS_1' (correlation with 'Pi_1' is 0.972145)\n",
      "         -> 2 statistics have been removed!\n",
      "- Performing an ABC-GLM standard estimations:\n",
      "   - Standardizing Parameters ... done!\n",
      "   - Standardizing statistics (distance calculation only)... done!\n",
      "   - Performing estimations for observed data set 0:\n",
      "      - Calculating distances ... done!\n",
      "      - Retaining the best 2000 simulations ... done!\n",
      "         -> 2000 simulations retained with tolerance 1.20775.\n",
      "      - Writing retained ... done!\n",
      "      - Calculating Tukey depth of observed data:\n",
      "         - Tukey depth = 0.262631\n",
      "      - Performing local linear regression ... done!\n",
      "      - Preparing posterior density calculation ... done!\n",
      "      - Calculating marginal posterior densities for 2 parameters ... done!\n",
      "      - Writing marginal posterior densities ... done!\n",
      "      - Writing posterior characteristics ... done!\n",
      "      - Computing joint posterior distribution(s):\n",
      "         - Computing joint posterior densities for parameters 'LOG10_MUTATION_RATE' and 'LOG10_N_NOW':\n",
      "            - Preparing grid of 100 equally spaced positions per parameter ...  done!\n",
      "            - Computing density at 10000 positions ... (0%)\r",
      "            - Computing density at 10000 positions ... (1%)\r",
      "            - Computing density at 10000 positions ... (2%)\r",
      "            - Computing density at 10000 positions ... (3%)\r",
      "            - Computing density at 10000 positions ... (4%)\r",
      "            - Computing density at 10000 positions ... (5%)\r",
      "            - Computing density at 10000 positions ... (6%)\r",
      "            - Computing density at 10000 positions ... (7%)\r",
      "            - Computing density at 10000 positions ... (8%)\r",
      "            - Computing density at 10000 positions ... (9%)\r",
      "            - Computing density at 10000 positions ... (10%)\r",
      "            - Computing density at 10000 positions ... (11%)\r",
      "            - Computing density at 10000 positions ... (12%)\r",
      "            - Computing density at 10000 positions ... (13%)\r",
      "            - Computing density at 10000 positions ... (14%)\r",
      "            - Computing density at 10000 positions ... (15%)\r",
      "            - Computing density at 10000 positions ... (16%)\r",
      "            - Computing density at 10000 positions ... (17%)\r",
      "            - Computing density at 10000 positions ... (18%)\r",
      "            - Computing density at 10000 positions ... (19%)\r",
      "            - Computing density at 10000 positions ... (20%)\r",
      "            - Computing density at 10000 positions ... (21%)\r",
      "            - Computing density at 10000 positions ... (22%)\r",
      "            - Computing density at 10000 positions ... (23%)\r",
      "            - Computing density at 10000 positions ... (24%)\r",
      "            - Computing density at 10000 positions ... (25%)\r",
      "            - Computing density at 10000 positions ... (26%)\r",
      "            - Computing density at 10000 positions ... (27%)\r",
      "            - Computing density at 10000 positions ... (28%)\r",
      "            - Computing density at 10000 positions ... (29%)\r",
      "            - Computing density at 10000 positions ... (30%)\r",
      "            - Computing density at 10000 positions ... (31%)\r",
      "            - Computing density at 10000 positions ... (32%)\r",
      "            - Computing density at 10000 positions ... (33%)\r",
      "            - Computing density at 10000 positions ... (34%)\r",
      "            - Computing density at 10000 positions ... (35%)\r",
      "            - Computing density at 10000 positions ... (36%)\r",
      "            - Computing density at 10000 positions ... (37%)\r",
      "            - Computing density at 10000 positions ... (38%)\r",
      "            - Computing density at 10000 positions ... (39%)\r",
      "            - Computing density at 10000 positions ... (40%)\r",
      "            - Computing density at 10000 positions ... (41%)\r",
      "            - Computing density at 10000 positions ... (42%)\r",
      "            - Computing density at 10000 positions ... (43%)\r",
      "            - Computing density at 10000 positions ... (44%)\r",
      "            - Computing density at 10000 positions ... (45%)\r",
      "            - Computing density at 10000 positions ... (46%)\r",
      "            - Computing density at 10000 positions ... (47%)\r",
      "            - Computing density at 10000 positions ... (48%)\r",
      "            - Computing density at 10000 positions ... (49%)\r",
      "            - Computing density at 10000 positions ... (50%)\r",
      "            - Computing density at 10000 positions ... (51%)\r",
      "            - Computing density at 10000 positions ... (52%)\r",
      "            - Computing density at 10000 positions ... (53%)\r",
      "            - Computing density at 10000 positions ... (54%)\r",
      "            - Computing density at 10000 positions ... (55%)\r",
      "            - Computing density at 10000 positions ... (56%)\r",
      "            - Computing density at 10000 positions ... (57%)\r",
      "            - Computing density at 10000 positions ... (58%)\r",
      "            - Computing density at 10000 positions ... (59%)\r",
      "            - Computing density at 10000 positions ... (60%)\r",
      "            - Computing density at 10000 positions ... (61%)\r",
      "            - Computing density at 10000 positions ... (62%)\r",
      "            - Computing density at 10000 positions ... (63%)\r",
      "            - Computing density at 10000 positions ... (64%)\r",
      "            - Computing density at 10000 positions ... (65%)\r",
      "            - Computing density at 10000 positions ... (66%)\r",
      "            - Computing density at 10000 positions ... (67%)\r",
      "            - Computing density at 10000 positions ... (68%)\r",
      "            - Computing density at 10000 positions ... (69%)\r",
      "            - Computing density at 10000 positions ... (70%)\r",
      "            - Computing density at 10000 positions ... (71%)\r",
      "            - Computing density at 10000 positions ... (72%)\r",
      "            - Computing density at 10000 positions ... (73%)\r",
      "            - Computing density at 10000 positions ... (74%)\r",
      "            - Computing density at 10000 positions ... (75%)\r",
      "            - Computing density at 10000 positions ... (76%)\r",
      "            - Computing density at 10000 positions ... (77%)\r",
      "            - Computing density at 10000 positions ... (78%)\r",
      "            - Computing density at 10000 positions ... (79%)\r",
      "            - Computing density at 10000 positions ... (80%)\r",
      "            - Computing density at 10000 positions ... (81%)\r",
      "            - Computing density at 10000 positions ... (82%)\r",
      "            - Computing density at 10000 positions ... (83%)\r",
      "            - Computing density at 10000 positions ... (84%)\r",
      "            - Computing density at 10000 positions ... (85%)\r",
      "            - Computing density at 10000 positions ... (86%)\r",
      "            - Computing density at 10000 positions ... (87%)\r",
      "            - Computing density at 10000 positions ... (88%)\r",
      "            - Computing density at 10000 positions ... (89%)\r",
      "            - Computing density at 10000 positions ... (90%)\r",
      "            - Computing density at 10000 positions ... (91%)\r",
      "            - Computing density at 10000 positions ... (92%)\r",
      "            - Computing density at 10000 positions ... (93%)\r",
      "            - Computing density at 10000 positions ... (94%)\r",
      "            - Computing density at 10000 positions ... (95%)\r",
      "            - Computing density at 10000 positions ... (96%)\r",
      "            - Computing density at 10000 positions ... (97%)\r",
      "            - Computing density at 10000 positions ... (98%)\r",
      "            - Computing density at 10000 positions ... (99%)\r",
      "            - Computing density at 10000 positions ... (100%)\r",
      "            - Computing density at 10000 positions ... done!        \n",
      "         - Writing file(s) with joint posterior estimates ... done!\n",
      "      - Writing model fit ... done!\n",
      "      - Estimations for data set 0 finished successfully!\n",
      "- Program terminated in 2.23333 min!\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: data/processed/: File exists\n",
      "mkdir: data/processed/simulated_observed_data: File exists\n",
      "ERROR: Output file 'sims_constsize_sampling1.txt' already exists! Use 'overWrite' to force replacement.\n",
      "\n",
      "mkdir: data/processed/abc_sims: File exists\n",
      "mkdir: data/processed/posterior: File exists\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ..\n",
    "sh commands.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<iframe src=../results/plots/2D_joint_posterior_n_u.pdf width=600 height=700></iframe>"
      ],
      "text/latex": [
       "\\includegraphics[width=1.0\\textwidth]{../results/plots/2D_joint_posterior_n_u.pdf}"
      ],
      "text/plain": [
       "<__main__.PDF at 0x10a3686d8>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class PDF(object):\n",
    "  def __init__(self, pdf, size=(200,200)):\n",
    "    self.pdf = pdf\n",
    "    self.size = size\n",
    "\n",
    "  def _repr_html_(self):\n",
    "    return '<iframe src={0} width={1[0]} height={1[1]}></iframe>'.format(self.pdf, self.size)\n",
    "\n",
    "  def _repr_latex_(self):\n",
    "    return r'\\includegraphics[width=1.0\\textwidth]{{{0}}}'.format(self.pdf)\n",
    "\n",
    "PDF('../results/plots/2D_joint_posterior_n_u.pdf',size=(600,700))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "[Overview](../main.ipynb) | [2. Expectation value](./expected_value.ipynb)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
